<!DOCTYPE html>
<html lang="en" dir="ltr" class="scroll-smooth" data-default-appearance="light"
  data-auto-appearance="true"><head>
  <meta charset="utf-8" />
  
  <meta http-equiv="content-language" content="en" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <meta http-equiv="X-UA-Compatible" content="ie=edge" />
  
  <title>利用blast对测序数据进行污染情况估计 &middot; </title>
  <meta name="title" content="利用blast对测序数据进行污染情况估计 &middot; " />
  
  <meta name="description" content="一个简单的小流程,利用blast来统计测序数据的污染情况。" />
  <meta name="keywords" content="guide, Bio, " />
  
  
  <link rel="canonical" href="https://with9.github.io/posts/blast_nt/" />
  
  
  
  
  
  
  
  
  
  
  <link type="text/css" rel="stylesheet" href="/css/main.bundle.min.00dd8b9350ecc82ac2af1f4674a9aad6c9a42e761652b3143d532e4b352bd5ca4750a9e040a76fd03e5d53ef1eb025fa8d60f7c673632f7ca0683ece7f990542.css"
    integrity="sha512-AN2Lk1DsyCrCrx9GdKmq1smkLnYWUrMUPVMuSzUr1cpHUKngQKdv0D5dU&#43;8esCX6jWD3xnNjL3ygaD7Of5kFQg==" />
  
  
  <script type="text/javascript" src="/js/appearance.min.516a16745bea5a9bd011138d254cc0fd3973cd55ce6e15f3dec763e7c7c2c7448f8fe7b54cca811cb821b0c7e12cd161caace1dd794ac3d34d40937cbcc9ee12.js"
    integrity="sha512-UWoWdFvqWpvQERONJUzA/TlzzVXObhXz3sdj58fCx0SPj&#43;e1TMqBHLghsMfhLNFhyqzh3XlKw9NNQJN8vMnuEg=="></script>
  
  
  
  
  
  
  
  
  
  
  
  
  <script defer type="text/javascript" id="script-bundle" src="/js/main.bundle.min.17ffe617e664d257760ee5f71869a6d9328d3613601c03a2eb7a194bbfb06ddc629a70f7b348e6e9446e81d0789555aebce09d27debd0329fc297601a1ee6a77.js"
    integrity="sha512-F//mF&#43;Zk0ld2DuX3GGmm2TKNNhNgHAOi63oZS7&#43;wbdximnD3s0jm6URugdB4lVWuvOCdJ969Ayn8KXYBoe5qdw==" data-copy="" data-copied=""></script>
  
  
  <script src="/js/zoom.min.js"></script>
  
  
  
  <link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png" />
  <link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png" />
  <link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png" />
  <link rel="manifest" href="/site.webmanifest" />
  
  
  
  
  
  
  
  <meta property="og:title" content="利用blast对测序数据进行污染情况估计" />
<meta property="og:description" content="一个简单的小流程,利用blast来统计测序数据的污染情况。" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://with9.github.io/posts/blast_nt/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2021-08-10T00:00:00+00:00" />
<meta property="article:modified_time" content="2021-08-10T00:00:00+00:00" />

  <meta name="twitter:card" content="summary"/><meta name="twitter:title" content="利用blast对测序数据进行污染情况估计"/>
<meta name="twitter:description" content="一个简单的小流程,利用blast来统计测序数据的污染情况。"/>

  
  <script type="application/ld+json">
  [{
    "@context": "https://schema.org",
    "@type": "Article",
    "articleSection": "Blog",
    "name": "利用blast对测序数据进行污染情况估计",
    "headline": "利用blast对测序数据进行污染情况估计",
    
    "abstract": "一个简单的小流程,利用blast来统计测序数据的污染情况。",
    "inLanguage": "en",
    "url" : "https:\/\/with9.github.io\/posts\/blast_nt\/",
    "author" : {
      "@type": "Person",
      "name": "with Φ"
    },
    "copyrightYear": "2021",
    "dateCreated": "2021-08-10T00:00:00\u002b00:00",
    "datePublished": "2021-08-10T00:00:00\u002b00:00",
    
    "dateModified": "2021-08-10T00:00:00\u002b00:00",
    
    "keywords": ["guide","Bio"],
    
    "mainEntityOfPage": "true",
    "wordCount": "900"
  }]
  </script>


  
  
  <meta name="author" content="with Φ" />
  
  
  
  <link href="https://github.com/with9" rel="me" />
  
  
  <link href="dc198424601@outlook.com" rel="me" />
  
  
  
  

<script src="/lib/jquery/jquery.slim.min.js" integrity=""></script>





















  
  


  
  
  <meta name="theme-color"/>
  
  
  
  <script src="https://www.gstatic.com/firebasejs/8.10.0/firebase-app.js"></script>
  <script src="https://www.gstatic.com/firebasejs/8.10.0/firebase-firestore.js"></script>
  <script src="https://www.gstatic.com/firebasejs/8.10.0/firebase-auth.js"></script>
  <script>

    const firebaseConfig = {
      apiKey: "AIzaSyCbyF5sSrwDj4NXHJwZ0VkWJtBLI4aIrwc",
      authDomain: "AIzaSyCbyF5sSrwDj4NXHJwZ0VkWJtBLI4aIrwc",
      projectId: "web-firebase-f8b18",
      storageBucket: "web-firebase-f8b18.appspot.com",
      messagingSenderId: "532704544245",
      appId: "1:532704544245:web:1b1f2d15d4fb7ff6c78cbc",
      measurementId: "G-0FJXQKW8XJ"
    };

    var app = firebase.initializeApp(firebaseConfig);
    var db = firebase.firestore();
    var auth = firebase.auth();

  </script>
  
  
</head>
<body
  class="flex flex-col h-screen px-6 m-auto text-lg leading-7 max-w-7xl bg-neutral text-neutral-900 dark:bg-neutral-800 dark:text-neutral sm:px-14 md:px-24 lg:px-32">
  <div id="the-top" class="absolute flex self-center">
    <a class="px-3 py-1 text-sm -translate-y-8 rounded-b-lg bg-primary-200 focus:translate-y-0 dark:bg-neutral-600"
      href="#main-content"><span
        class="font-bold text-primary-600 ltr:pr-2 rtl:pl-2 dark:text-primary-400">&darr;</span>Skip to main content</a>
  </div>
  
  
  <div class="min-h-[148px]"></div>
<div class="fixed inset-x-0 pl-[24px] pr-[24px]" style="z-index:100">
  <div id="menu-blur" class="absolute opacity-0 inset-x-0 top-0 h-full single_hero_background nozoom backdrop-blur-2xl shadow-2xl"></div>
  <div class="relative max-w-[64rem] ml-auto mr-auto">
    <div style="padding-left:0;padding-right:0;padding-top:2px;padding-bottom:3px"
    class="main-menu flex items-center justify-between px-4 py-6 sm:px-6 md:justify-start space-x-3">
    
    
    
    <div>
        <a href="/" class="flex">
            <span class="sr-only"></span>

            <img src="/img/logo.png" width="80" height="18"
                class="logo max-h-[5rem] max-w-[5rem] object-scale-down object-left nozoom" alt="" />

        </a>
    </div>
    
    <div class="flex flex-1 items-center justify-between">
        <nav class="flex space-x-3">

            
            <a href="/" class="text-base font-medium text-gray-500 hover:text-gray-900"></a>
            

        </nav>
        <nav class="hidden md:flex items-center space-x-5 md:ml-12 h-12">

            
            
            
  <a href="/posts/"  class="flex items-center text-gray-500 hover:text-primary-600 dark:hover:text-primary-400">
    
    <p class="text-base font-medium" title="">
        Blog
    </p>
</a>


            
            
  <a href="/about/"  class="flex items-center text-gray-500 hover:text-primary-600 dark:hover:text-primary-400">
    
    <p class="text-base font-medium" title="">
        About
    </p>
</a>


            
            
  <a href="/friends/"  class="flex items-center text-gray-500 hover:text-primary-600 dark:hover:text-primary-400">
    
    <p class="text-base font-medium" title="">
        Friends
    </p>
</a>


            
            
  <a href="/posts/read/"  class="flex items-center text-gray-500 hover:text-primary-600 dark:hover:text-primary-400">
    
    <p class="text-base font-medium" title="">
        Read
    </p>
</a>


            
            
  <a href="https://github.com/with9"  target="_blank"  class="flex items-center text-gray-500 hover:text-primary-600 dark:hover:text-primary-400">
    
    <span >
        

  <span class="relative block icon">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path fill="currentColor" d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>

  </span>


    </span>
    
    <p class="text-base font-medium" title="">
        
    </p>
</a>


            
            

            


            
            <button id="search-button" aria-label="Search" class="text-base hover:text-primary-600 dark:hover:text-primary-400"
                title="">
                

  <span class="relative block icon">
    <svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="search" class="svg-inline--fa fa-search fa-w-16" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path fill="currentColor" d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"/></svg>

  </span>


            </button>
            


            
            
            <div
                class="ltr:mr-14 rtl:ml-14 flex items-center">
                <button id="appearance-switcher" aria-label="Dark mode switcher" type="button" class="text-base hover:text-primary-600 dark:hover:text-primary-400">
                    <div class="flex items-center justify-center dark:hidden">
                        

  <span class="relative block icon">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path fill="currentColor" d="M32 256c0-123.8 100.3-224 223.8-224c11.36 0 29.7 1.668 40.9 3.746c9.616 1.777 11.75 14.63 3.279 19.44C245 86.5 211.2 144.6 211.2 207.8c0 109.7 99.71 193 208.3 172.3c9.561-1.805 16.28 9.324 10.11 16.95C387.9 448.6 324.8 480 255.8 480C132.1 480 32 379.6 32 256z"/></svg>

  </span>


                    </div>
                    <div class="items-center justify-center hidden dark:flex">
                        

  <span class="relative block icon">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path fill="currentColor" d="M256 159.1c-53.02 0-95.1 42.98-95.1 95.1S202.1 351.1 256 351.1s95.1-42.98 95.1-95.1S309 159.1 256 159.1zM509.3 347L446.1 255.1l63.15-91.01c6.332-9.125 1.104-21.74-9.826-23.72l-109-19.7l-19.7-109c-1.975-10.93-14.59-16.16-23.72-9.824L256 65.89L164.1 2.736c-9.125-6.332-21.74-1.107-23.72 9.824L121.6 121.6L12.56 141.3C1.633 143.2-3.596 155.9 2.736 164.1L65.89 256l-63.15 91.01c-6.332 9.125-1.105 21.74 9.824 23.72l109 19.7l19.7 109c1.975 10.93 14.59 16.16 23.72 9.824L256 446.1l91.01 63.15c9.127 6.334 21.75 1.107 23.72-9.822l19.7-109l109-19.7C510.4 368.8 515.6 356.1 509.3 347zM256 383.1c-70.69 0-127.1-57.31-127.1-127.1c0-70.69 57.31-127.1 127.1-127.1s127.1 57.3 127.1 127.1C383.1 326.7 326.7 383.1 256 383.1z"/></svg>

  </span>


                    </div>
                </button>
            </div>
            

        </nav>
        <div class="flex md:hidden items-center space-x-5 md:ml-12 h-12">

            <span></span>

            


            
            <button id="search-button-mobile" aria-label="Search" class="text-base hover:text-primary-600 dark:hover:text-primary-400"
                title="">
                

  <span class="relative block icon">
    <svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="search" class="svg-inline--fa fa-search fa-w-16" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path fill="currentColor" d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"/></svg>

  </span>


            </button>
            

            
            
            <button id="appearance-switcher-mobile" aria-label="Dark mode switcher" type="button" class="text-base hover:text-primary-600 dark:hover:text-primary-400" style="margin-right:5px">
                <div class="flex items-center justify-center dark:hidden">
                    

  <span class="relative block icon">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path fill="currentColor" d="M32 256c0-123.8 100.3-224 223.8-224c11.36 0 29.7 1.668 40.9 3.746c9.616 1.777 11.75 14.63 3.279 19.44C245 86.5 211.2 144.6 211.2 207.8c0 109.7 99.71 193 208.3 172.3c9.561-1.805 16.28 9.324 10.11 16.95C387.9 448.6 324.8 480 255.8 480C132.1 480 32 379.6 32 256z"/></svg>

  </span>


                </div>
                <div class="items-center justify-center hidden dark:flex">
                    

  <span class="relative block icon">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path fill="currentColor" d="M256 159.1c-53.02 0-95.1 42.98-95.1 95.1S202.1 351.1 256 351.1s95.1-42.98 95.1-95.1S309 159.1 256 159.1zM509.3 347L446.1 255.1l63.15-91.01c6.332-9.125 1.104-21.74-9.826-23.72l-109-19.7l-19.7-109c-1.975-10.93-14.59-16.16-23.72-9.824L256 65.89L164.1 2.736c-9.125-6.332-21.74-1.107-23.72 9.824L121.6 121.6L12.56 141.3C1.633 143.2-3.596 155.9 2.736 164.1L65.89 256l-63.15 91.01c-6.332 9.125-1.105 21.74 9.824 23.72l109 19.7l19.7 109c1.975 10.93 14.59 16.16 23.72 9.824L256 446.1l91.01 63.15c9.127 6.334 21.75 1.107 23.72-9.822l19.7-109l109-19.7C510.4 368.8 515.6 356.1 509.3 347zM256 383.1c-70.69 0-127.1-57.31-127.1-127.1c0-70.69 57.31-127.1 127.1-127.1s127.1 57.3 127.1 127.1C383.1 326.7 326.7 383.1 256 383.1z"/></svg>

  </span>


                </div>
            </button>
            

        </div>
    </div>
    <div class="-my-2 -mr-2 md:hidden">

        <label id="menu-button" for="menu-controller" class="block">
            <input type="checkbox" id="menu-controller" class="hidden" />
            
            <div class="cursor-pointer hover:text-primary-600 dark:hover:text-primary-400">
                

  <span class="relative block icon">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path fill="currentColor" d="M0 96C0 78.33 14.33 64 32 64H416C433.7 64 448 78.33 448 96C448 113.7 433.7 128 416 128H32C14.33 128 0 113.7 0 96zM0 256C0 238.3 14.33 224 32 224H416C433.7 224 448 238.3 448 256C448 273.7 433.7 288 416 288H32C14.33 288 0 273.7 0 256zM416 448H32C14.33 448 0 433.7 0 416C0 398.3 14.33 384 32 384H416C433.7 384 448 398.3 448 416C448 433.7 433.7 448 416 448z"/></svg>

  </span>


            </div>
            <div id="menu-wrapper" style="padding-top:5px;"
                class="fixed inset-0 z-30 invisible w-screen h-screen m-0 overflow-auto transition-opacity opacity-0 cursor-default bg-neutral-100/50 backdrop-blur-sm dark:bg-neutral-900/50">
                <ul
                    class="flex space-y-2 mt-3 flex-col items-end w-full px-6 py-6 mx-auto overflow-visible list-none ltr:text-right rtl:text-left max-w-7xl">

                    <li>
                        <span
                            class="cursor-pointer inline-block align-text-bottom hover:text-primary-600 dark:hover:text-primary-400">

  <span class="relative block icon">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 320 512"><path fill="currentColor" d="M310.6 361.4c12.5 12.5 12.5 32.75 0 45.25C304.4 412.9 296.2 416 288 416s-16.38-3.125-22.62-9.375L160 301.3L54.63 406.6C48.38 412.9 40.19 416 32 416S15.63 412.9 9.375 406.6c-12.5-12.5-12.5-32.75 0-45.25l105.4-105.4L9.375 150.6c-12.5-12.5-12.5-32.75 0-45.25s32.75-12.5 45.25 0L160 210.8l105.4-105.4c12.5-12.5 32.75-12.5 45.25 0s12.5 32.75 0 45.25l-105.4 105.4L310.6 361.4z"/></svg>

  </span>

</span>
                    </li>

                    

                    
  <li class="mt-1">
    <a href="/posts/"  class="flex items-center text-gray-500 hover:text-primary-600 dark:hover:text-primary-400">
        
        <p class="text-bg font-bg" title="">
            Blog
        </p>
    </a>
</li>



                    

                    
  <li class="mt-1">
    <a href="/about/"  class="flex items-center text-gray-500 hover:text-primary-600 dark:hover:text-primary-400">
        
        <p class="text-bg font-bg" title="">
            About
        </p>
    </a>
</li>



                    

                    
  <li class="mt-1">
    <a href="/friends/"  class="flex items-center text-gray-500 hover:text-primary-600 dark:hover:text-primary-400">
        
        <p class="text-bg font-bg" title="">
            Friends
        </p>
    </a>
</li>



                    

                    
  <li class="mt-1">
    <a href="/posts/read/"  class="flex items-center text-gray-500 hover:text-primary-600 dark:hover:text-primary-400">
        
        <p class="text-bg font-bg" title="">
            Read
        </p>
    </a>
</li>



                    

                    
  <li class="mt-1">
    <a href="https://github.com/with9"  target="_blank"  class="flex items-center text-gray-500 hover:text-primary-600 dark:hover:text-primary-400">
        
        <div >
            

  <span class="relative block icon">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path fill="currentColor" d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>

  </span>


        </div>
        
        <p class="text-bg font-bg" title="">
            
        </p>
    </a>
</li>



                    

                </ul>
                
                

            </div>
        </label>
    </div>
</div>




<script>
    (function () {
        var $mainmenu = $('.main-menu');
        var path = window.location.pathname;
        $mainmenu.find('a[href="' + path + '"]').each(function (i, e) {
            $(e).children('p').addClass('active');
        });
    })();
</script>


  </div>
</div>
<script>
  window.addEventListener('scroll', function (e) {
    var scroll = window.pageYOffset || document.documentElement.scrollTop || document.body.scrollTop || 0;
    var background_blur = document.getElementById('menu-blur');
    background_blur.style.opacity = (scroll / 300);
  });
</script>

  
  <div class="relative flex flex-col grow">
    <main id="main-content" class="grow">
      


<article>
  
  
  
  
  
  
 
  
   
  
 


        
        <figure>
            <img class="w-full rounded-lg single_hero_round nozoom" alt="利用blast对测序数据进行污染情况估计" src="/img/background.svg">
            
        </figure>
        
    
  
  

  <header id="single_header" class="mt-5 max-w-prose">
    
    <ol class="text-sm text-neutral-500 dark:text-neutral-400 print:hidden">
  
  
    
  
    
  
  <li class="inline hidden">
    <a
      class="hover:underline decoration-neutral-300 dark:underline-neutral-600"
      href="/"
      >About</a
    ><span class="px-1 text-primary-500">/</span>
  </li>

  
  <li class="inline ">
    <a
      class="hover:underline decoration-neutral-300 dark:underline-neutral-600"
      href="/posts/"
      >Blog</a
    ><span class="px-1 text-primary-500">/</span>
  </li>

  
  <li class="inline hidden">
    <a
      class="hover:underline decoration-neutral-300 dark:underline-neutral-600"
      href="/posts/blast_nt/"
      >利用blast对测序数据进行污染情况估计</a
    ><span class="px-1 text-primary-500">/</span>
  </li>

</ol>


    
    <h1 class="mt-0 text-4xl font-extrabold text-neutral-900 dark:text-neutral">
      利用blast对测序数据进行污染情况估计
    </h1>
    <div class="mt-1 mb-6 text-base text-neutral-500 dark:text-neutral-400 print:hidden">
      





  
  







  









  









  



<div class="flex flex-row flex-wrap items-center">
  
  
  <time datetime="2021-08-10 00:00:00 &#43;0000 UTC">2021-08-10</time><span class="px-2 text-primary-500">&middot;</span><span>
    
        
        
            
            
        
    <span id="views_posts/blast_nt/index.md" 
    
        class="animate-pulse inline-block text-transparent max-h-3 rounded-full mt-[-2px] align-middle bg-neutral-300 dark:bg-neutral-400" title="views">loading</span> 
    <span class="inline-block align-text-bottom">

  <span class="relative block icon">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 576 512">
<path fill="currentColor" d="M288 32c-80.8 0-145.5 36.8-192.6 80.6C48.6 156 17.3 208 2.5 243.7c-3.3 7.9-3.3 16.7 0 24.6C17.3 304 48.6 356 95.4 399.4C142.5 443.2 207.2 480 288 480s145.5-36.8 192.6-80.6c46.8-43.5 78.1-95.4 93-131.1c3.3-7.9 3.3-16.7 0-24.6c-14.9-35.7-46.2-87.7-93-131.1C433.5 68.8 368.8 32 288 32zM432 256c0 79.5-64.5 144-144 144s-144-64.5-144-144s64.5-144 144-144s144 64.5 144 144zM288 192c0 35.3-28.7 64-64 64c-11.5 0-22.3-3-31.6-8.4c-.2 2.8-.4 5.5-.4 8.4c0 53 43 96 96 96s96-43 96-96s-43-96-96-96c-2.8 0-5.6 .1-8.4 .4c5.3 9.3 8.4 20.1 8.4 31.6z"/></svg>
  </span>

</span>
</span><span class="px-2 text-primary-500">&middot;</span>


<script type="text/javascript" src="/js/zen-mode.min.ccc8b2a0c4505216444ca25530c7328aa0bef94f43dabaf3e598e1b4271de9eec76c1cb35917471a4f7ccb8d84051b3acef42baa30724ac3a667fcf1fec4432a.js" integrity="sha512-zMiyoMRQUhZETKJVMMcyiqC&#43;&#43;U9D2rrz5ZjhtCcd6e7HbByzWRdHGk98y42EBRs6zvQrqjBySsOmZ/zx/sRDKg=="></script>

<span class="mb-[2px]">
    <span id="zen-mode-button"
          class="text-lg hover:text-primary-500"
          title="Enable zen mode"
          data-title-i18n-disable="Enable zen mode"
          data-title-i18n-enable="Disable zen mode">
        <span class="inline-block align-text-bottom">

  <span class="relative block icon">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 50 50" width="50px" height="50px">
    <path fill="currentColor" d="M 12.980469 4 C 9.1204688 4 5.9804688 7.14 5.9804688 11 L 6 26 L 9.9804688 26 L 9.9804688 11 C 9.9804688 9.35 11.320469 8 12.980469 8 L 40.019531 8 C 41.679531 8 43.019531 9.35 43.019531 11 L 43.019531 39 C 43.019531 40.65 41.679531 42 40.019531 42 L 29 42 C 29 43.54 28.420938 44.94 27.460938 46 L 40.019531 46 C 43.879531 46 47.019531 42.86 47.019531 39 L 47.019531 11 C 47.019531 7.14 43.879531 4 40.019531 4 L 12.980469 4 z M 7 28 C 4.794 28 3 29.794 3 32 L 3 42 C 3 44.206 4.794 46 7 46 L 23 46 C 25.206 46 27 44.206 27 42 L 27 32 C 27 29.794 25.206 28 23 28 L 7 28 z M 7 32 L 23 32 L 23.001953 42 L 7 42 L 7 32 z"/>
</svg>
  </span>

</span>
    </span>
</span>
  

  
  
</div>


<div class="flex flex-row flex-wrap items-center">
  
  
  
  
  
  
  
  
  
  
  
</div>




<div class="flex flex-row flex-wrap items-center">
  
  
  
  
  
  
  <span style="margin-top:0.5rem" class="mr-2" onclick="window.open(&#34;/categories/bio/&#34;,'_self');">
    <span class="flex" style="cursor: pointer;">
  <span class="rounded-md border border-primary-400 px-1 py-[1px] text-xs font-normal text-primary-700 dark:border-primary-600 dark:text-primary-400">
    Bio
  </span>
</span>
  </span>
  
  
  
  
  
  
  
  
  
  <span style="margin-top:0.5rem" class="mr-2" onclick="window.open(&#34;/tags/guide/&#34;,'_self');">
    <span class="flex" style="cursor: pointer;">
  <span class="rounded-md border border-primary-400 px-1 py-[1px] text-xs font-normal text-primary-700 dark:border-primary-600 dark:text-primary-400">
    Guide
  </span>
</span>
  </span>
  
  <span style="margin-top:0.5rem" class="mr-2" onclick="window.open(&#34;/tags/bio/&#34;,'_self');">
    <span class="flex" style="cursor: pointer;">
  <span class="rounded-md border border-primary-400 px-1 py-[1px] text-xs font-normal text-primary-700 dark:border-primary-600 dark:text-primary-400">
    Bio
  </span>
</span>
  </span>
  
  
  
  
</div>



    </div>

    
    
    
    
    

    

    

    

    
  
  </header>
  
  <section class="flex flex-col max-w-full mt-0 prose dark:prose-invert lg:flex-row">
    
     <div
      class="order-first lg:ml-auto px-0 lg:order-last ltr:lg:pl-8 rtl:lg:pr-8">
      <div class="toc ltr:pl-5 rtl:pr-5 print:hidden lg:sticky lg:top-[140px]">

         <details open class="toc-right mt-0 overflow-hidden rounded-lg ltr:-ml-5 ltr:pl-5 rtl:-mr-5 rtl:pr-5 hidden lg:block">
  <summary
    class="block py-1 text-lg font-semibold cursor-pointer bg-neutral-100 text-neutral-800 ltr:-ml-5 ltr:pl-5 rtl:-mr-5 rtl:pr-5 dark:bg-neutral-700 dark:text-neutral-100 lg:hidden">
    Table of Contents
  </summary>
  <div
    class="min-w-[220px] py-2 border-dotted border-neutral-300 ltr:-ml-5 ltr:border-l ltr:pl-5 rtl:-mr-5 rtl:border-r rtl:pr-5 dark:border-neutral-600">
    <nav id="TableOfContents">
  <ul>
    <li><a href="#前期数据库下载及预处理">前期数据库下载及预处理</a>
      <ul>
        <li><a href="#nt库下载">nt库下载</a></li>
        <li><a href="#物种名与基因id对应数据库下载">物种名与基因id对应数据库下载</a></li>
      </ul>
    </li>
    <li><a href="#对测序数据随机抽取并blast">对测序数据随机抽取并blast</a></li>
    <li><a href="#snakemake脚本">snakemake脚本</a></li>
  </ul>
</nav>
  </div>
</details>
<details class="toc-inside mt-0 overflow-hidden rounded-lg ltr:-ml-5 ltr:pl-5 rtl:-mr-5 rtl:pr-5 lg:hidden">
  <summary
    class="py-1 text-lg font-semibold cursor-pointer bg-neutral-100 text-neutral-800 ltr:-ml-5 ltr:pl-5 rtl:-mr-5 rtl:pr-5 dark:bg-neutral-700 dark:text-neutral-100 lg:hidden">
    Table of Contents
  </summary>
  <div
    class="py-2 border-dotted border-neutral-300 ltr:-ml-5 ltr:border-l ltr:pl-5 rtl:-mr-5 rtl:border-r rtl:pr-5 dark:border-neutral-600">
    <nav id="TableOfContents">
  <ul>
    <li><a href="#前期数据库下载及预处理">前期数据库下载及预处理</a>
      <ul>
        <li><a href="#nt库下载">nt库下载</a></li>
        <li><a href="#物种名与基因id对应数据库下载">物种名与基因id对应数据库下载</a></li>
      </ul>
    </li>
    <li><a href="#对测序数据随机抽取并blast">对测序数据随机抽取并blast</a></li>
    <li><a href="#snakemake脚本">snakemake脚本</a></li>
  </ul>
</nav>
  </div>
</details>

 
<script>
  (function () {
    var $toc = $('#TableOfContents');
    if ($toc.length > 0) {
      var $window = $(window);

      function onScroll() {
        var currentScroll = $window.scrollTop();
        var h = $('.anchor');
        var id = "";
        h.each(function (i, e) {
          e = $(e);
          if (e.offset().top - $(window).height()/3 <= currentScroll) {
            id = e.attr('id');
          }
        });
        var active = $toc.find('a.active');      
        if (active.length == 1 && active.eq(0).attr('href') == '#' + id) return true;

        active.each(function (i, e) {
          
            $(e).removeClass('active');
          
        });
        $toc.find('a[href="#' + id + '"]').addClass('active')
        $toc.find('a[href="#' + id + '"]').parentsUntil('#TableOfContents').each(function (i, e) {
          $(e).children('a').parents('ul').show();          
        });
      }

      $window.on('scroll', onScroll);
      $(document).ready(function () {
        
        onScroll();
      });
    }
  })();
</script>
   </div>
      </div>
      

      <div class="min-w-0 min-h-0 max-w-fit">
        
        


        <div class="article-content max-w-prose mb-20">
          <p>一般二代调研中需要判断一下测序样品的污染情况，主要的原理就是对样品的测序数据随机抽取并进行blast，分析reads的来源情况，从而判断样品污染情况。</p>
<p>主要流程参考<a href="https://blog.csdn.net/qq_42962326/article/details/105081327"   target="_blank">
    这篇博客</a></p>


<h2 class="relative group">前期数据库下载及预处理 
    <div id="%E5%89%8D%E6%9C%9F%E6%95%B0%E6%8D%AE%E5%BA%93%E4%B8%8B%E8%BD%BD%E5%8F%8A%E9%A2%84%E5%A4%84%E7%90%86" class="anchor"></div>
    
    <span
        class="absolute top-0 w-6 transition-opacity opacity-0 ltr:-left-6 rtl:-right-6 not-prose group-hover:opacity-100">
        <a class="group-hover:text-primary-300 dark:group-hover:text-neutral-700"
            style="text-decoration-line: none !important;" href="#%E5%89%8D%E6%9C%9F%E6%95%B0%E6%8D%AE%E5%BA%93%E4%B8%8B%E8%BD%BD%E5%8F%8A%E9%A2%84%E5%A4%84%E7%90%86" aria-label="Anchor">#</a>
    </span>        
    
</h2>


<h3 class="relative group">nt库下载 
    <div id="nt%E5%BA%93%E4%B8%8B%E8%BD%BD" class="anchor"></div>
    
    <span
        class="absolute top-0 w-6 transition-opacity opacity-0 ltr:-left-6 rtl:-right-6 not-prose group-hover:opacity-100">
        <a class="group-hover:text-primary-300 dark:group-hover:text-neutral-700"
            style="text-decoration-line: none !important;" href="#nt%E5%BA%93%E4%B8%8B%E8%BD%BD" aria-label="Anchor">#</a>
    </span>        
    
</h3>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="c1">#nt.41.tar.gz</span>
</span></span><span class="line"><span class="cl"><span class="c1">#https://ftp.ncbi.nlm.nih.gov/blast/db/nt.20.tar.gz</span>
</span></span><span class="line"><span class="cl"><span class="c1">#----down.py</span>
</span></span><span class="line"><span class="cl"><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">42</span><span class="p">):</span>
</span></span><span class="line"><span class="cl">    <span class="k">if</span> <span class="n">i</span> <span class="o">&lt;</span><span class="mi">10</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">        <span class="n">url</span><span class="o">=</span><span class="s2">&#34;https://ftp.ncbi.nlm.nih.gov/blast/db/nt.</span><span class="si">{}{}</span><span class="s2">.tar.gz&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;0&#39;</span><span class="p">,</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
</span></span><span class="line"><span class="cl">        <span class="nb">print</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">    <span class="k">else</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">        <span class="n">url</span><span class="o">=</span><span class="s2">&#34;https://ftp.ncbi.nlm.nih.gov/blast/db/nt.</span><span class="si">{}</span><span class="s2">.tar.gz&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
</span></span><span class="line"><span class="cl">        <span class="nb">print</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-shell" data-lang="shell"><span class="line"><span class="cl">python down.py&gt;down
</span></span><span class="line"><span class="cl">aria2c --input<span class="o">=</span>down <span class="c1">#进行下载</span>
</span></span><span class="line"><span class="cl"><span class="c1">#分别解压</span>
</span></span><span class="line"><span class="cl">mkdir all
</span></span><span class="line"><span class="cl">ls nt*/* <span class="p">|</span>xargs -i ln <span class="o">{}</span> all/ <span class="c1">#建立硬链接到同一个目录</span>
</span></span></code></pre></div>

<h3 class="relative group">物种名与基因id对应数据库下载 
    <div id="%E7%89%A9%E7%A7%8D%E5%90%8D%E4%B8%8E%E5%9F%BA%E5%9B%A0id%E5%AF%B9%E5%BA%94%E6%95%B0%E6%8D%AE%E5%BA%93%E4%B8%8B%E8%BD%BD" class="anchor"></div>
    
    <span
        class="absolute top-0 w-6 transition-opacity opacity-0 ltr:-left-6 rtl:-right-6 not-prose group-hover:opacity-100">
        <a class="group-hover:text-primary-300 dark:group-hover:text-neutral-700"
            style="text-decoration-line: none !important;" href="#%E7%89%A9%E7%A7%8D%E5%90%8D%E4%B8%8E%E5%9F%BA%E5%9B%A0id%E5%AF%B9%E5%BA%94%E6%95%B0%E6%8D%AE%E5%BA%93%E4%B8%8B%E8%BD%BD" aria-label="Anchor">#</a>
    </span>        
    
</h3>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-shell" data-lang="shell"><span class="line"><span class="cl">aria2c <span class="s1">&#39;https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/accession2taxid/nucl_gb.accession2taxid.gz&#39;</span><span class="c1">#基因id对应taxonomy id</span>
</span></span><span class="line"><span class="cl">aria2c <span class="s1">&#39;https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz&#39;</span><span class="c1">#taxonomy id 对应学名</span>
</span></span><span class="line"><span class="cl"><span class="c1">#分别解压</span>
</span></span><span class="line"><span class="cl"><span class="c1">#---head nucl_gb.accession2taxid</span>
</span></span><span class="line"><span class="cl">accession	accession.version	taxid	gi
</span></span><span class="line"><span class="cl">A00001	A00001.1	10641	<span class="m">58418</span>
</span></span><span class="line"><span class="cl">A00002	A00002.1	9913	<span class="m">2</span>
</span></span><span class="line"><span class="cl">A00003	A00003.1	9913	<span class="m">3</span>
</span></span><span class="line"><span class="cl">A00004	A00004.1	32630	<span class="m">57971</span>
</span></span><span class="line"><span class="cl">A00005	A00005.1	32630	<span class="m">57972</span>
</span></span><span class="line"><span class="cl">A00006	A00006.1	32630	<span class="m">57973</span>
</span></span><span class="line"><span class="cl">A00008	A00008.1	32630	<span class="m">57974</span>
</span></span><span class="line"><span class="cl">A00009	A00009.1	32630	<span class="m">57975</span>
</span></span><span class="line"><span class="cl">A00010	A00010.1	32630	<span class="m">57976</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1">#---head taxdump/names.dmp</span>
</span></span><span class="line"><span class="cl">1	<span class="p">|</span>	all	<span class="p">|</span>		<span class="p">|</span>	synonym	<span class="p">|</span>
</span></span><span class="line"><span class="cl">1	<span class="p">|</span>	root	<span class="p">|</span>		<span class="p">|</span>	scientific name	<span class="p">|</span>
</span></span><span class="line"><span class="cl">2	<span class="p">|</span>	Bacteria	<span class="p">|</span>	Bacteria &lt;bacteria&gt;	<span class="p">|</span>	scientific name	<span class="p">|</span>
</span></span><span class="line"><span class="cl">2	<span class="p">|</span>	bacteria	<span class="p">|</span>		<span class="p">|</span>	blast name	<span class="p">|</span>
</span></span><span class="line"><span class="cl">2	<span class="p">|</span>	eubacteria	<span class="p">|</span>		<span class="p">|</span>	genbank common name	<span class="p">|</span>
</span></span><span class="line"><span class="cl">2	<span class="p">|</span>	Monera	<span class="p">|</span>	Monera &lt;bacteria&gt;	<span class="p">|</span>	in-part	<span class="p">|</span>
</span></span><span class="line"><span class="cl">2	<span class="p">|</span>	Procaryotae	<span class="p">|</span>	Procaryotae &lt;bacteria&gt;	<span class="p">|</span>	in-part	<span class="p">|</span>
</span></span><span class="line"><span class="cl">2	<span class="p">|</span>	Prokaryotae	<span class="p">|</span>	Prokaryotae &lt;bacteria&gt;	<span class="p">|</span>	in-part	<span class="p">|</span>
</span></span><span class="line"><span class="cl">2	<span class="p">|</span>	Prokaryota	<span class="p">|</span>	Prokaryota &lt;bacteria&gt;	<span class="p">|</span>	in-part	<span class="p">|</span>
</span></span><span class="line"><span class="cl">2	<span class="p">|</span>	prokaryote	<span class="p">|</span>	prokaryote &lt;bacteria&gt;	<span class="p">|</span>	in-part	<span class="p">|</span>
</span></span><span class="line"><span class="cl">2	<span class="p">|</span>	prokaryotes	<span class="p">|</span>	prokaryotes &lt;bacteria&gt;	<span class="p">|</span>	in-part	<span class="p">|</span>
</span></span><span class="line"><span class="cl">6	<span class="p">|</span>	Azorhizobium Dreyfus et al. <span class="m">1988</span> emend. Lang et al. 2013	<span class="p">|</span>		<span class="p">|</span>	authority	<span class="p">|</span>
</span></span><span class="line"><span class="cl">6	<span class="p">|</span>	Azorhizobium	<span class="p">|</span>		<span class="p">|</span>	scientific name	<span class="p">|</span>
</span></span><span class="line"><span class="cl">7	<span class="p">|</span>	ATCC 43989	<span class="p">|</span>	ATCC <span class="m">43989</span> &lt;<span class="nb">type</span> strain&gt;	<span class="p">|</span>	<span class="nb">type</span> material	<span class="p">|</span>
</span></span><span class="line"><span class="cl">7	<span class="p">|</span>	Azorhizobium caulinodans Dreyfus et al. 1988	<span class="p">|</span>		<span class="p">|</span>	authority	<span class="p">|</span>
</span></span><span class="line"><span class="cl">7	<span class="p">|</span>	Azorhizobium caulinodans	<span class="p">|</span>		<span class="p">|</span>	scientific name	<span class="p">|</span>
</span></span><span class="line"><span class="cl">7	<span class="p">|</span>	Azotirhizobium caulinodans	<span class="p">|</span>		<span class="p">|</span>	equivalent name	<span class="p">|</span>
</span></span><span class="line"><span class="cl">7	<span class="p">|</span>	CCUG 26647	<span class="p">|</span>	CCUG <span class="m">26647</span> &lt;<span class="nb">type</span> strain&gt;	<span class="p">|</span>	<span class="nb">type</span> material	<span class="p">|</span>
</span></span><span class="line"><span class="cl">7	<span class="p">|</span>	DSM 5975	<span class="p">|</span>	DSM <span class="m">5975</span> &lt;<span class="nb">type</span> strain&gt;	<span class="p">|</span>	<span class="nb">type</span> material	<span class="p">|</span>
</span></span><span class="line"><span class="cl">7	<span class="p">|</span>	IFO 14845	<span class="p">|</span>	IFO <span class="m">14845</span> &lt;<span class="nb">type</span> strain&gt;	<span class="p">|</span>	<span class="nb">type</span> material	<span class="p">|</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">rga <span class="s2">&#34;scientific name&#34;</span> -g <span class="s2">&#34;names.dmp&#34;</span> &gt;names_scientific#获取scientific name相应的行
</span></span></code></pre></div>

<h2 class="relative group">对测序数据随机抽取并blast 
    <div id="%E5%AF%B9%E6%B5%8B%E5%BA%8F%E6%95%B0%E6%8D%AE%E9%9A%8F%E6%9C%BA%E6%8A%BD%E5%8F%96%E5%B9%B6blast" class="anchor"></div>
    
    <span
        class="absolute top-0 w-6 transition-opacity opacity-0 ltr:-left-6 rtl:-right-6 not-prose group-hover:opacity-100">
        <a class="group-hover:text-primary-300 dark:group-hover:text-neutral-700"
            style="text-decoration-line: none !important;" href="#%E5%AF%B9%E6%B5%8B%E5%BA%8F%E6%95%B0%E6%8D%AE%E9%9A%8F%E6%9C%BA%E6%8A%BD%E5%8F%96%E5%B9%B6blast" aria-label="Anchor">#</a>
    </span>        
    
</h2>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-shell" data-lang="shell"><span class="line"><span class="cl"><span class="c1">#从fastq文件中抽取5000条数据</span>
</span></span><span class="line"><span class="cl">cat 1.fastq <span class="p">|</span> head -n <span class="m">20000</span> <span class="p">|</span> awk <span class="s1">&#39;{if(NR%4==1){print &#34;&gt;&#34;$1}else if(NR%4==2){print $0}}&#39;</span><span class="p">|</span>sed <span class="s1">&#39;s/@//g&#39;</span> &gt; myfile.fa
</span></span><span class="line"><span class="cl"><span class="c1">#---head myfile.fa</span>
</span></span><span class="line"><span class="cl">&gt;SRR15225403.1
</span></span><span class="line"><span class="cl">TTATCGTAGATTTTTTGAATATGTTGTAACGAAAGCTGAGCCATGTTGATTCCTTTGCTATACGTTTTGCCAGAAGGCCGATAACCGCGGCCAGGTGAGTGATTCCGTTGAAGGCAACAGTAATTGCGCCCCGGTTAAGCCCGCGCCGATCCCCACCGAGCGCATACCGCTGGCGTTAATGGCGTCAATGCCCGCCTGCGCATCTTCAATGCCGATACATGCCTGCGGCGGCAC
</span></span><span class="line"><span class="cl">&gt;SRR15225403.2
</span></span><span class="line"><span class="cl">GAAGGTTACTGCGGCTCCTGTCGCACCCGTCTGGTTGCAGGTCAGGTTGACTGGATTGCCGAACCGTTAGCCTTTATTCAGCCGGGGGAAATTTTGCCCTGTTGTTG
</span></span><span class="line"><span class="cl">&gt;SRR15225403.3
</span></span><span class="line"><span class="cl">CGGTAAACGCCGTCAGGGAGATGGCGCAACCAATCGCCAACGGCAGATTAGCCCACAGACCCATCACGATAGAACCGAGTCCGGCAACCAGACAGGTCGCAACGAAAACTGCCGCAGGCGGGAAGCCCGCTTTACCCAACATGCCTGGAACGACGATAACCGAGTAGACCATCGCCAGAAACGTTGTTAACCCGGCAACCACTTCCTGACGGACAGTGCTTCCACGTTGTGAAAT
</span></span><span class="line"><span class="cl">&gt;SRR15225403.4
</span></span><span class="line"><span class="cl">CAGGCGAAAGCCGCGGCGGTAAAAGTCCACGCTGACGCGCCCGGTACGTTTTATTGCGGATGTAAAATTAACTGGCAGGGCAAAAAAGGCGTTGTTGATCTGCAATCGTGCGGCTATCAGGTGCGCAAAAATGAAAACCGAGCCAGCCGCGTAGAGTGGGAACACGTCGTTCCCGCCTGGCAGTTCGGTCACCAGCGCCAGTGCTGGCAGGACGGGGGACGTAAAAACTGCGCTA
</span></span><span class="line"><span class="cl">&gt;SRR15225403.5
</span></span><span class="line"><span class="cl">TGTATCTGAATCCGCAAGATTGCAGCGTGATTAATGATGAAGCGCTGAATCGTATTATCGCCGTAGGCCACCAGCATCATCTGAACGTTGTCGGCTGGAACCCGGGACCGGCGCTTTCAGTTAGCATGGGCGACATGCCGGATGATGGCTACAAACCATTGGTTGGGTAGAAAAGGCCTACGGCTTCGGAACGGCAAAAGGGAACAAAAGGGAAACTGCCACACCTGGCGAATC
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1">#进行blastn 并将结果保存为xml格式</span>
</span></span><span class="line"><span class="cl">blastn -query myfile.fa -out xml/out00.xml -max_target_seqs <span class="m">1</span> -outfmt <span class="m">5</span> -db ~/usb/blastn/all/nt -num_threads <span class="m">8</span> -evalue 1e-5
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1">#---head out00.xml</span>
</span></span><span class="line"><span class="cl">&lt;?xml <span class="nv">version</span><span class="o">=</span><span class="s2">&#34;1.0&#34;</span>?&gt;
</span></span><span class="line"><span class="cl">&lt;!DOCTYPE BlastOutput PUBLIC <span class="s2">&#34;-//NCBI//NCBI BlastOutput/EN&#34;</span> <span class="s2">&#34;http://www.ncbi.nlm.nih.gov/dtd/NCBI_BlastOutput.dtd&#34;</span>&gt;
</span></span><span class="line"><span class="cl">&lt;BlastOutput&gt;
</span></span><span class="line"><span class="cl">  &lt;BlastOutput_program&gt;blastn&lt;/BlastOutput_program&gt;
</span></span><span class="line"><span class="cl">  &lt;BlastOutput_version&gt;BLASTN 2.12.0+&lt;/BlastOutput_version&gt;
</span></span><span class="line"><span class="cl">  &lt;BlastOutput_reference&gt;Zheng Zhang, Scott Schwartz, Lukas Wagner, and Webb Miller <span class="o">(</span>2000<span class="o">)</span>, <span class="p">&amp;</span>quot<span class="p">;</span>A greedy algorithm <span class="k">for</span> aligning DNA sequences<span class="p">&amp;</span>quot<span class="p">;</span>, J Comput Biol 2000<span class="p">;</span> 7<span class="o">(</span>1-2<span class="o">)</span>:203-14.&lt;/BlastOutput_reference&gt;
</span></span><span class="line"><span class="cl">  &lt;BlastOutput_db&gt;/home/with9/usb/blastn/nt.00/nt.00&lt;/BlastOutput_db&gt;
</span></span><span class="line"><span class="cl">  &lt;BlastOutput_query-ID&gt;Query_1&lt;/BlastOutput_query-ID&gt;
</span></span><span class="line"><span class="cl">  &lt;BlastOutput_query-def&gt;SRR15225403.1&lt;/BlastOutput_query-def&gt;
</span></span><span class="line"><span class="cl">  &lt;BlastOutput_query-len&gt;234&lt;/BlastOutput_query-len&gt;
</span></span></code></pre></div><p>编写python脚本处理out.xml文件,将gid,hit_def和hit_accession写入csv文件</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="kn">import</span> <span class="nn">xml.etree.ElementTree</span> <span class="k">as</span> <span class="nn">ET</span>
</span></span><span class="line"><span class="cl"><span class="kn">import</span> <span class="nn">csv</span>
</span></span><span class="line"><span class="cl"><span class="kn">import</span> <span class="nn">os</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="k">def</span> <span class="nf">readonexml</span><span class="p">(</span><span class="n">fid</span><span class="p">:</span><span class="nb">int</span><span class="p">):</span>
</span></span><span class="line"><span class="cl">    <span class="k">if</span> <span class="n">fid</span><span class="o">&lt;</span><span class="mi">10</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">        <span class="n">xmlfile</span><span class="o">=</span><span class="s2">&#34;xml/out0</span><span class="si">{}</span><span class="s2">.xml&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">fid</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">        <span class="n">csvfile</span><span class="o">=</span><span class="s2">&#34;csv/out0</span><span class="si">{}</span><span class="s2">.csv&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">fid</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">    <span class="k">else</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">        <span class="n">xmlfile</span><span class="o">=</span><span class="s2">&#34;xml/out</span><span class="si">{}</span><span class="s2">.xml&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">fid</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">        <span class="n">csvfile</span><span class="o">=</span><span class="s2">&#34;csv/out</span><span class="si">{}</span><span class="s2">.csv&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">fid</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">    <span class="n">tree</span> <span class="o">=</span> <span class="n">ET</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">xmlfile</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">    <span class="n">root</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">getroot</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">    <span class="n">hits</span><span class="o">=</span><span class="n">root</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="s1">&#39;.//Hit&#39;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">csvfile</span><span class="p">,</span><span class="s1">&#39;w&#39;</span><span class="p">,</span><span class="n">newline</span><span class="o">=</span><span class="s2">&#34;&#34;</span><span class="p">)</span><span class="k">as</span> <span class="n">f</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">        <span class="n">csv_writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">        <span class="n">csv_writer</span><span class="o">.</span><span class="n">writerow</span><span class="p">([</span><span class="s2">&#34;Hit_id&#34;</span><span class="p">,</span><span class="s2">&#34;Hit_def&#34;</span><span class="p">,</span><span class="s2">&#34;Hit_accession&#34;</span><span class="p">])</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">        <span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">hits</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">            <span class="n">hit_id</span><span class="o">=</span><span class="n">hit</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;Hit_id&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>
</span></span><span class="line"><span class="cl">            <span class="n">hit_def</span><span class="o">=</span><span class="n">hit</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;Hit_def&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>
</span></span><span class="line"><span class="cl">            <span class="n">hit_accession</span><span class="o">=</span><span class="n">hit</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;Hit_accession&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>
</span></span><span class="line"><span class="cl">            <span class="n">csv_writer</span><span class="o">.</span><span class="n">writerow</span><span class="p">([</span><span class="n">hit_id</span><span class="p">,</span><span class="n">hit_def</span><span class="p">,</span><span class="n">hit_accession</span><span class="p">])</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">readonexml</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1">#--head csv/out00.csv</span>
</span></span><span class="line"><span class="cl"><span class="n">Hit_id</span><span class="p">,</span><span class="n">Hit_def</span><span class="p">,</span><span class="n">Hit_accession</span>
</span></span><span class="line"><span class="cl"><span class="n">gi</span><span class="o">|</span><span class="mi">85674274</span><span class="o">|</span><span class="n">dbj</span><span class="o">|</span><span class="n">AP009048</span><span class="mf">.1</span><span class="o">|</span><span class="p">,</span><span class="s2">&#34;Escherichia coli str. K-12 substr. W3110 DNA, complete genome&#34;</span><span class="p">,</span><span class="n">AP009048</span>
</span></span><span class="line"><span class="cl"><span class="n">gi</span><span class="o">|</span><span class="mi">85674274</span><span class="o">|</span><span class="n">dbj</span><span class="o">|</span><span class="n">AP009048</span><span class="mf">.1</span><span class="o">|</span><span class="p">,</span><span class="s2">&#34;Escherichia coli str. K-12 substr. W3110 DNA, complete genome&#34;</span><span class="p">,</span><span class="n">AP009048</span>
</span></span><span class="line"><span class="cl"><span class="n">gi</span><span class="o">|</span><span class="mi">81239530</span><span class="o">|</span><span class="n">gb</span><span class="o">|</span><span class="n">CP000034</span><span class="mf">.1</span><span class="o">|</span><span class="p">,</span><span class="s2">&#34;Shigella dysenteriae Sd197, complete genome&#34;</span><span class="p">,</span><span class="n">CP000034</span>
</span></span><span class="line"><span class="cl"><span class="n">gi</span><span class="o">|</span><span class="mi">81244029</span><span class="o">|</span><span class="n">gb</span><span class="o">|</span><span class="n">CP000036</span><span class="mf">.1</span><span class="o">|</span><span class="p">,</span><span class="s2">&#34;Shigella boydii Sb227, complete genome&#34;</span><span class="p">,</span><span class="n">CP000036</span>
</span></span><span class="line"><span class="cl"><span class="n">gi</span><span class="o">|</span><span class="mi">30043918</span><span class="o">|</span><span class="n">gb</span><span class="o">|</span><span class="n">AE014073</span><span class="mf">.1</span><span class="o">|</span><span class="p">,</span><span class="s2">&#34;Shigella flexneri 2a str. 2457T, complete genome&#34;</span><span class="p">,</span><span class="n">AE014073</span>
</span></span><span class="line"><span class="cl"><span class="n">gi</span><span class="o">|</span><span class="mi">30043918</span><span class="o">|</span><span class="n">gb</span><span class="o">|</span><span class="n">AE014073</span><span class="mf">.1</span><span class="o">|</span><span class="p">,</span><span class="s2">&#34;Shigella flexneri 2a str. 2457T, complete genome&#34;</span><span class="p">,</span><span class="n">AE014073</span>
</span></span><span class="line"><span class="cl"><span class="n">gi</span><span class="o">|</span><span class="mi">81244029</span><span class="o">|</span><span class="n">gb</span><span class="o">|</span><span class="n">CP000036</span><span class="mf">.1</span><span class="o">|</span><span class="p">,</span><span class="s2">&#34;Shigella boydii Sb227, complete genome&#34;</span><span class="p">,</span><span class="n">CP000036</span>
</span></span><span class="line"><span class="cl"><span class="n">gi</span><span class="o">|</span><span class="mi">85674274</span><span class="o">|</span><span class="n">dbj</span><span class="o">|</span><span class="n">AP009048</span><span class="mf">.1</span><span class="o">|</span><span class="p">,</span><span class="s2">&#34;Escherichia coli str. K-12 substr. W3110 DNA, complete genome&#34;</span><span class="p">,</span><span class="n">AP009048</span>
</span></span><span class="line"><span class="cl"><span class="n">gi</span><span class="o">|</span><span class="mi">30043918</span><span class="o">|</span><span class="n">gb</span><span class="o">|</span><span class="n">AE014073</span><span class="mf">.1</span><span class="o">|</span><span class="p">,</span><span class="s2">&#34;Shigella flexneri 2a str. 2457T, complete genome&#34;</span><span class="p">,</span><span class="n">AE014073</span>
</span></span></code></pre></div><p>利用Hit_accession从nucl_gb.accession2taxid中获取taxonomyid,因为文件过大,所以利用os模块调用<a href="https://github.com/phiresky/ripgrep-all"   target="_blank">
    rga</a>工具进行查找</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="c1">#统计出所有需要查找的Hit_accession,因为文件过大,所以利用os模块调用rga工具进行查找</span>
</span></span><span class="line"><span class="cl"><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
</span></span><span class="line"><span class="cl"><span class="kn">import</span> <span class="nn">os</span>
</span></span><span class="line"><span class="cl"><span class="n">df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;csv/out00.csv&#39;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;Hitaccession.csv&#39;</span><span class="p">,</span><span class="s1">&#39;w&#39;</span><span class="p">)</span><span class="k">as</span> <span class="n">f</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">    <span class="n">name_df</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Hit_accession&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">    <span class="n">name_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">    
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;Hitacession.csv&#39;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">df</span><span class="p">[</span><span class="s1">&#39;tax_id&#39;</span><span class="p">]</span><span class="o">=</span><span class="mi">0</span>
</span></span><span class="line"><span class="cl"><span class="n">tids</span><span class="o">=</span><span class="p">[]</span>
</span></span><span class="line"><span class="cl"><span class="k">for</span> <span class="n">index</span><span class="p">,</span><span class="n">row</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
</span></span><span class="line"><span class="cl">    <span class="n">hit_accession</span><span class="o">=</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</span></span><span class="line"><span class="cl">    <span class="n">cmd</span><span class="o">=</span><span class="s2">&#34;rga &#39;&#34;</span><span class="o">+</span><span class="n">hit_accession</span><span class="o">+</span><span class="s2">&#34;&#39; -g &#39;nucl_gb.accession2taxid.gz&#39;|awk &#39;{print$3}&#39;&#34;</span>
</span></span><span class="line"><span class="cl">    <span class="n">tid</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">popen</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
</span></span><span class="line"><span class="cl">    <span class="nb">print</span><span class="p">(</span><span class="n">tid</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">    <span class="n">tids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tid</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">df</span><span class="p">[</span><span class="s1">&#39;tax_id&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">tids</span>
</span></span><span class="line"><span class="cl"><span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;Tid_result,csv&#39;</span><span class="p">)</span>
</span></span></code></pre></div><p>对一开始的读取xml脚本进行修改,同时获取物种id和物种的学名.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="kn">import</span> <span class="nn">xml.etree.ElementTree</span> <span class="k">as</span> <span class="nn">ET</span>
</span></span><span class="line"><span class="cl"><span class="kn">import</span> <span class="nn">csv</span>
</span></span><span class="line"><span class="cl"><span class="kn">import</span> <span class="nn">os</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="k">def</span> <span class="nf">readonexml</span><span class="p">(</span><span class="n">fid</span><span class="p">:</span><span class="nb">int</span><span class="p">):</span>
</span></span><span class="line"><span class="cl">    <span class="k">if</span> <span class="n">fid</span><span class="o">&lt;</span><span class="mi">10</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">        <span class="n">xmlfile</span><span class="o">=</span><span class="s2">&#34;xml/out0</span><span class="si">{}</span><span class="s2">.xml&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">fid</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">        <span class="n">csvfile</span><span class="o">=</span><span class="s2">&#34;csv/out0</span><span class="si">{}</span><span class="s2">.csv&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">fid</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">    <span class="k">else</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">        <span class="n">xmlfile</span><span class="o">=</span><span class="s2">&#34;xml/out</span><span class="si">{}</span><span class="s2">.xml&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">fid</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">        <span class="n">csvfile</span><span class="o">=</span><span class="s2">&#34;csv/out</span><span class="si">{}</span><span class="s2">.csv&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">fid</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">    <span class="n">tree</span> <span class="o">=</span> <span class="n">ET</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">xmlfile</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">    <span class="n">root</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">getroot</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">    <span class="n">hits</span><span class="o">=</span><span class="n">root</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="s1">&#39;.//Hit&#39;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">csvfile</span><span class="p">,</span><span class="s1">&#39;w&#39;</span><span class="p">,</span><span class="n">newline</span><span class="o">=</span><span class="s2">&#34;&#34;</span><span class="p">)</span><span class="k">as</span> <span class="n">f</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">        <span class="n">csv_writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">        <span class="n">csv_writer</span><span class="o">.</span><span class="n">writerow</span><span class="p">([</span><span class="s2">&#34;Hit_id&#34;</span><span class="p">,</span><span class="s2">&#34;Hit_def&#34;</span><span class="p">,</span><span class="s2">&#34;Hit_accession&#34;</span><span class="p">,</span><span class="s2">&#34;taxonomy_id&#34;</span><span class="p">,</span><span class="s2">&#34;scientific_name&#34;</span><span class="p">])</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">        <span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">hits</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">            <span class="n">hit_id</span><span class="o">=</span><span class="n">hit</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;Hit_id&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>
</span></span><span class="line"><span class="cl">            <span class="n">hit_def</span><span class="o">=</span><span class="n">hit</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;Hit_def&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>
</span></span><span class="line"><span class="cl">            <span class="n">hit_accession</span><span class="o">=</span><span class="n">hit</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;Hit_accession&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>
</span></span><span class="line"><span class="cl">            <span class="n">cmd</span><span class="o">=</span><span class="s2">&#34;grep &#39;</span><span class="si">{}</span><span class="s2">&#39; Tid_result.csv |cut -d , -f 4&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">hit_accession</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">            <span class="n">taxonomy_id</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">popen</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
</span></span><span class="line"><span class="cl">            <span class="nb">print</span><span class="p">(</span><span class="n">taxonomy_id</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">            <span class="n">cmd2</span><span class="o">=</span><span class="s2">&#34;grep &#39;:</span><span class="si">{}</span><span class="s2">&#39; names_scientific |cut -d \| -f 2&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">taxonomy_id</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">            <span class="n">scientific_name</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">popen</span><span class="p">(</span><span class="n">cmd2</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
</span></span><span class="line"><span class="cl">            <span class="n">csv_writer</span><span class="o">.</span><span class="n">writerow</span><span class="p">([</span><span class="n">hit_id</span><span class="p">,</span><span class="n">hit_def</span><span class="p">,</span><span class="n">hit_accession</span><span class="p">,</span><span class="n">taxonomy_id</span><span class="p">,</span><span class="n">scientific_name</span><span class="p">])</span>
</span></span><span class="line"><span class="cl"><span class="n">readonexml</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
</span></span></code></pre></div><p>统计结果并绘图</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;csv/out00.csv&#39;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;temp_with_name.csv&#39;</span><span class="p">,</span><span class="s1">&#39;w&#39;</span><span class="p">)</span><span class="k">as</span> <span class="n">f</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">    <span class="n">name_df</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;scientific_name&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">    <span class="n">name_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">name_df</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">pie</span><span class="p">()</span>
</span></span><span class="line"><span class="cl"><span class="n">df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;temp_with_name.csv&#39;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">df</span><span class="p">[</span><span class="s1">&#39;points&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;scientific_name&#39;</span><span class="p">]</span><span class="o">/</span><span class="mi">5000</span><span class="o">*</span><span class="mi">100</span>
</span></span><span class="line"><span class="cl"><span class="n">df</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">set_axis</span><span class="p">([</span><span class="s2">&#34;scientific_name&#34;</span><span class="p">,</span><span class="s2">&#34;hits&#34;</span><span class="p">,</span><span class="s2">&#34;points&#34;</span><span class="p">],</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&#34;point.csv&#34;</span><span class="p">,</span><span class="s2">&#34;w&#34;</span><span class="p">)</span><span class="k">as</span> <span class="n">f</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">    <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="p">,</span><span class="n">float_format</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%.3f</span><span class="s1">&#39;</span><span class="p">,</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="c1">#---head point.csv</span>
</span></span><span class="line"><span class="cl"><span class="n">scientific_name</span><span class="p">,</span><span class="n">hits</span><span class="p">,</span><span class="n">points</span>
</span></span><span class="line"><span class="cl"><span class="n">Escherichia</span> <span class="n">coli</span> <span class="nb">str</span><span class="o">.</span> <span class="n">K</span><span class="o">-</span><span class="mi">12</span> <span class="n">substr</span><span class="o">.</span> <span class="n">W3110</span><span class="p">,</span><span class="mi">1785</span><span class="p">,</span><span class="mf">35.700</span>
</span></span><span class="line"><span class="cl"><span class="n">Shigella</span> <span class="n">boydii</span> <span class="n">Sb227</span><span class="p">,</span><span class="mi">668</span><span class="p">,</span><span class="mf">13.360</span>
</span></span><span class="line"><span class="cl"><span class="n">Shigella</span> <span class="n">flexneri</span> <span class="mi">2</span><span class="n">a</span> <span class="nb">str</span><span class="o">.</span> <span class="mi">2457</span><span class="n">T</span><span class="p">,</span><span class="mi">614</span><span class="p">,</span><span class="mf">12.280</span>
</span></span><span class="line"><span class="cl"><span class="n">Shigella</span> <span class="n">sonnei</span> <span class="n">Ss046</span><span class="p">,</span><span class="mi">438</span><span class="p">,</span><span class="mf">8.760</span>
</span></span><span class="line"><span class="cl"><span class="n">Escherichia</span> <span class="n">coli</span><span class="p">,</span><span class="mi">244</span><span class="p">,</span><span class="mf">4.880</span>
</span></span><span class="line"><span class="cl"><span class="n">Escherichia</span> <span class="n">coli</span> <span class="n">CFT073</span><span class="p">,</span><span class="mi">242</span><span class="p">,</span><span class="mf">4.840</span>
</span></span><span class="line"><span class="cl"><span class="n">Escherichia</span> <span class="n">coli</span> <span class="n">O157</span><span class="p">:</span><span class="n">H7</span> <span class="nb">str</span><span class="o">.</span> <span class="n">EDL933</span><span class="p">,</span><span class="mi">233</span><span class="p">,</span><span class="mf">4.660</span>
</span></span><span class="line"><span class="cl"><span class="n">Shigella</span> <span class="n">dysenteriae</span> <span class="n">Sd197</span><span class="p">,</span><span class="mi">149</span><span class="p">,</span><span class="mf">2.980</span>
</span></span><span class="line"><span class="cl"><span class="n">Escherichia</span> <span class="n">coli</span> <span class="n">K</span><span class="o">-</span><span class="mi">12</span><span class="p">,</span><span class="mi">55</span><span class="p">,</span><span class="mf">1.100</span>
</span></span></code></pre></div>

<h2 class="relative group">snakemake脚本 
    <div id="snakemake%E8%84%9A%E6%9C%AC" class="anchor"></div>
    
    <span
        class="absolute top-0 w-6 transition-opacity opacity-0 ltr:-left-6 rtl:-right-6 not-prose group-hover:opacity-100">
        <a class="group-hover:text-primary-300 dark:group-hover:text-neutral-700"
            style="text-decoration-line: none !important;" href="#snakemake%E8%84%9A%E6%9C%AC" aria-label="Anchor">#</a>
    </span>        
    
</h2>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="line"><span class="cl"><span class="c1">#输入工作环境</span>
</span></span><span class="line"><span class="cl"><span class="o">~/</span><span class="n">D</span><span class="o">/</span><span class="n">t</span><span class="o">/</span><span class="n">o</span><span class="o">/</span><span class="n">snake_test</span><span class="err">➜</span> <span class="n">tree</span> <span class="o">./</span>
</span></span><span class="line"><span class="cl"><span class="o">./</span>
</span></span><span class="line"><span class="cl"><span class="err">├──</span> <span class="n">data</span>
</span></span><span class="line"><span class="cl"><span class="err">│</span>   <span class="err">└──</span> <span class="mf">1.</span><span class="n">fastq</span>
</span></span><span class="line"><span class="cl"><span class="err">├──</span> <span class="n">names_scientific</span>
</span></span><span class="line"><span class="cl"><span class="err">├──</span> <span class="n">nucl_gb</span><span class="o">.</span><span class="n">accession2taxid</span><span class="o">.</span><span class="n">gz</span>
</span></span><span class="line"><span class="cl"><span class="err">└──</span> <span class="n">snakefile</span>
</span></span><span class="line"><span class="cl"><span class="o">--------------------------------------------------------------------------------------------</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">rule</span> <span class="nb">all</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">    <span class="nb">input</span><span class="p">:</span> <span class="s2">&#34;out/points.csv&#34;</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">rule</span> <span class="n">get_sample</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">    <span class="nb">input</span><span class="p">:</span> <span class="s2">&#34;data/1.fastq&#34;</span>
</span></span><span class="line"><span class="cl">    <span class="n">output</span><span class="p">:</span> <span class="s2">&#34;data/myfile.fa&#34;</span>
</span></span><span class="line"><span class="cl">    <span class="n">shell</span><span class="p">:</span> <span class="s2">&#34;cat </span><span class="si">{input}</span><span class="s2"> | head -n 20000 | awk &#39;{{if(NR%4==1){{print </span><span class="se">\&#34;</span><span class="s2">&gt;</span><span class="se">\&#34;</span><span class="s2">$1}}else if(NR%4==2){{print $0}}}}&#39;|sed &#39;s/@//&#39; &gt; </span><span class="si">{output}</span><span class="s2">&#34;</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">rule</span> <span class="n">to_blast</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">    <span class="nb">input</span><span class="p">:</span> <span class="n">rules</span><span class="o">.</span><span class="n">get_sample</span><span class="o">.</span><span class="n">output</span>
</span></span><span class="line"><span class="cl">    <span class="n">output</span><span class="p">:</span> <span class="s2">&#34;out/out.xml&#34;</span>
</span></span><span class="line"><span class="cl">    <span class="n">shell</span><span class="p">:</span> <span class="s2">&#34;blastn -query </span><span class="si">{input}</span><span class="s2"> -out </span><span class="si">{output}</span><span class="s2"> -max_target_seqs 1 -outfmt 5 -db ~/usb/blastn/all/nt -num_threads 8 -evalue 1e-5&#34;</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">rule</span> <span class="n">one_csv</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">    <span class="nb">input</span><span class="p">:</span>  <span class="s2">&#34;out/out.xml&#34;</span>
</span></span><span class="line"><span class="cl">    <span class="n">output</span><span class="p">:</span> <span class="s2">&#34;out/one_csv.csv&#34;</span> 
</span></span><span class="line"><span class="cl">    <span class="n">run</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">        <span class="kn">import</span> <span class="nn">xml.etree.ElementTree</span> <span class="k">as</span> <span class="nn">ET</span>
</span></span><span class="line"><span class="cl">        <span class="kn">import</span> <span class="nn">csv</span>
</span></span><span class="line"><span class="cl">        <span class="kn">import</span> <span class="nn">os</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">        <span class="k">def</span> <span class="nf">readonexml</span><span class="p">(</span><span class="n">fid</span><span class="p">:</span><span class="nb">int</span><span class="p">):</span>
</span></span><span class="line"><span class="cl">            <span class="c1"># if fid&lt;10:</span>
</span></span><span class="line"><span class="cl">            <span class="c1">#     xmlfile=&#34;xml/out0{}.xml&#34;.format(fid)</span>
</span></span><span class="line"><span class="cl">            <span class="c1">#     csvfile=&#34;csv/out0{}.csv&#34;.format(fid)</span>
</span></span><span class="line"><span class="cl">            <span class="c1"># else:</span>
</span></span><span class="line"><span class="cl">            <span class="c1">#     xmlfile=&#34;xml/out{}.xml&#34;.format(fid)</span>
</span></span><span class="line"><span class="cl">            <span class="c1">#     csvfile=&#34;csv/out{}.csv&#34;.format(fid)</span>
</span></span><span class="line"><span class="cl">            <span class="n">xmlfile</span><span class="o">=</span><span class="s2">&#34;out/out.xml&#34;</span>
</span></span><span class="line"><span class="cl">            <span class="n">csvfile</span><span class="o">=</span><span class="s2">&#34;out/one_csv.csv&#34;</span>
</span></span><span class="line"><span class="cl">            <span class="n">tree</span> <span class="o">=</span> <span class="n">ET</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">xmlfile</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">            <span class="n">root</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">getroot</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">            <span class="n">hits</span><span class="o">=</span><span class="n">root</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="s1">&#39;.//Hit&#39;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">csvfile</span><span class="p">,</span><span class="s1">&#39;w&#39;</span><span class="p">,</span><span class="n">newline</span><span class="o">=</span><span class="s2">&#34;&#34;</span><span class="p">)</span><span class="k">as</span> <span class="n">f</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">                <span class="n">csv_writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">                <span class="n">csv_writer</span><span class="o">.</span><span class="n">writerow</span><span class="p">([</span><span class="s2">&#34;Hit_id&#34;</span><span class="p">,</span><span class="s2">&#34;Hit_def&#34;</span><span class="p">,</span><span class="s2">&#34;Hit_accession&#34;</span><span class="p">])</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">                <span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">hits</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">                    <span class="n">hit_id</span><span class="o">=</span><span class="n">hit</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;Hit_id&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>
</span></span><span class="line"><span class="cl">                    <span class="n">hit_def</span><span class="o">=</span><span class="n">hit</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;Hit_def&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>
</span></span><span class="line"><span class="cl">                    <span class="n">hit_accession</span><span class="o">=</span><span class="n">hit</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;Hit_accession&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>
</span></span><span class="line"><span class="cl">                    <span class="n">csv_writer</span><span class="o">.</span><span class="n">writerow</span><span class="p">([</span><span class="n">hit_id</span><span class="p">,</span><span class="n">hit_def</span><span class="p">,</span><span class="n">hit_accession</span><span class="p">])</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">        <span class="n">readonexml</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">rule</span> <span class="n">find_id</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">    <span class="nb">input</span><span class="p">:</span> <span class="s2">&#34;out/one_csv.csv&#34;</span>
</span></span><span class="line"><span class="cl">    <span class="n">output</span><span class="p">:</span> <span class="s2">&#34;out/Tid_result.csv&#34;</span>
</span></span><span class="line"><span class="cl">    <span class="n">run</span><span class="p">:</span> 
</span></span><span class="line"><span class="cl">        <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
</span></span><span class="line"><span class="cl">        <span class="kn">import</span> <span class="nn">os</span>
</span></span><span class="line"><span class="cl">        <span class="n">df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;out/one_csv.csv&#39;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;Hitacession.csv&#39;</span><span class="p">,</span><span class="s1">&#39;w&#39;</span><span class="p">)</span><span class="k">as</span> <span class="n">f</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">            <span class="n">name_df</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Hit_accession&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">            <span class="n">name_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">            
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">        <span class="n">df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;Hitacession.csv&#39;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;tax_id&#39;</span><span class="p">]</span><span class="o">=</span><span class="mi">0</span>
</span></span><span class="line"><span class="cl">        <span class="n">tids</span><span class="o">=</span><span class="p">[]</span>
</span></span><span class="line"><span class="cl">        <span class="k">for</span> <span class="n">index</span><span class="p">,</span><span class="n">row</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
</span></span><span class="line"><span class="cl">            <span class="n">hit_accession</span><span class="o">=</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</span></span><span class="line"><span class="cl">            <span class="n">cmd</span><span class="o">=</span><span class="s2">&#34;rga &#39;&#34;</span><span class="o">+</span><span class="n">hit_accession</span><span class="o">+</span><span class="s2">&#34;&#39; -g &#39;nucl_gb.accession2taxid.gz&#39;|awk &#39;{print$3}&#39;&#34;</span>
</span></span><span class="line"><span class="cl">            <span class="n">tid</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">popen</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
</span></span><span class="line"><span class="cl">            <span class="nb">print</span><span class="p">(</span><span class="n">tid</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">            <span class="n">tids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tid</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;tax_id&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">tids</span>
</span></span><span class="line"><span class="cl">        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;out/Tid_result.csv&#39;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">rule</span> <span class="n">two_csv</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">    <span class="nb">input</span><span class="p">:</span> <span class="s2">&#34;out/Tid_result.csv&#34;</span>
</span></span><span class="line"><span class="cl">    <span class="n">output</span><span class="p">:</span> <span class="s2">&#34;out/two_csv.csv&#34;</span>
</span></span><span class="line"><span class="cl">    <span class="n">run</span><span class="p">:</span> 
</span></span><span class="line"><span class="cl">        <span class="kn">import</span> <span class="nn">xml.etree.ElementTree</span> <span class="k">as</span> <span class="nn">ET</span>
</span></span><span class="line"><span class="cl">        <span class="kn">import</span> <span class="nn">csv</span>
</span></span><span class="line"><span class="cl">        <span class="kn">import</span> <span class="nn">os</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">        <span class="k">def</span> <span class="nf">readonexml</span><span class="p">(</span><span class="n">fid</span><span class="p">:</span><span class="nb">int</span><span class="p">):</span>
</span></span><span class="line"><span class="cl">            <span class="c1"># if fid&lt;10:</span>
</span></span><span class="line"><span class="cl">            <span class="c1">#     xmlfile=&#34;xml/out0{}.xml&#34;.format(fid)</span>
</span></span><span class="line"><span class="cl">            <span class="c1">#     csvfile=&#34;csv/out0{}.csv&#34;.format(fid)</span>
</span></span><span class="line"><span class="cl">            <span class="c1"># else:</span>
</span></span><span class="line"><span class="cl">            <span class="c1">#     xmlfile=&#34;xml/out{}.xml&#34;.format(fid)</span>
</span></span><span class="line"><span class="cl">            <span class="c1">#     csvfile=&#34;csv/out{}.csv&#34;.format(fid)</span>
</span></span><span class="line"><span class="cl">            <span class="n">xmlfile</span><span class="o">=</span><span class="s2">&#34;out/out.xml&#34;</span>
</span></span><span class="line"><span class="cl">            <span class="n">csvfile</span><span class="o">=</span><span class="s2">&#34;out/two_csv.csv&#34;</span>
</span></span><span class="line"><span class="cl">            <span class="n">tree</span> <span class="o">=</span> <span class="n">ET</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">xmlfile</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">            <span class="n">root</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">getroot</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">            <span class="n">hits</span><span class="o">=</span><span class="n">root</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="s1">&#39;.//Hit&#39;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">csvfile</span><span class="p">,</span><span class="s1">&#39;w&#39;</span><span class="p">,</span><span class="n">newline</span><span class="o">=</span><span class="s2">&#34;&#34;</span><span class="p">)</span><span class="k">as</span> <span class="n">f</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">                <span class="n">csv_writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">                <span class="n">csv_writer</span><span class="o">.</span><span class="n">writerow</span><span class="p">([</span><span class="s2">&#34;Hit_id&#34;</span><span class="p">,</span><span class="s2">&#34;Hit_def&#34;</span><span class="p">,</span><span class="s2">&#34;Hit_accession&#34;</span><span class="p">,</span><span class="s2">&#34;taxonomy_id&#34;</span><span class="p">,</span><span class="s2">&#34;scientific_name&#34;</span><span class="p">])</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">                <span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">hits</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">                    <span class="n">hit_id</span><span class="o">=</span><span class="n">hit</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;Hit_id&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>
</span></span><span class="line"><span class="cl">                    <span class="n">hit_def</span><span class="o">=</span><span class="n">hit</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;Hit_def&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>
</span></span><span class="line"><span class="cl">                    <span class="n">hit_accession</span><span class="o">=</span><span class="n">hit</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;Hit_accession&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>
</span></span><span class="line"><span class="cl">                    <span class="n">cmd</span><span class="o">=</span><span class="s2">&#34;grep &#39;</span><span class="si">{}</span><span class="s2">&#39; out/Tid_result.csv |cut -d , -f 4&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">hit_accession</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">                    <span class="n">taxonomy_id</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">popen</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
</span></span><span class="line"><span class="cl">                    <span class="nb">print</span><span class="p">(</span><span class="n">taxonomy_id</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">                    <span class="n">cmd2</span><span class="o">=</span><span class="s2">&#34;grep &#39;:</span><span class="si">{}</span><span class="s2">&#39; names_scientific |cut -d \| -f 2&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">taxonomy_id</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">                    <span class="n">scientific_name</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">popen</span><span class="p">(</span><span class="n">cmd2</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
</span></span><span class="line"><span class="cl">                    <span class="n">csv_writer</span><span class="o">.</span><span class="n">writerow</span><span class="p">([</span><span class="n">hit_id</span><span class="p">,</span><span class="n">hit_def</span><span class="p">,</span><span class="n">hit_accession</span><span class="p">,</span><span class="n">taxonomy_id</span><span class="p">,</span><span class="n">scientific_name</span><span class="p">])</span>
</span></span><span class="line"><span class="cl">        <span class="n">readonexml</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">rule</span> <span class="n">get_points</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">    <span class="nb">input</span><span class="p">:</span> <span class="s2">&#34;out/two_csv.csv&#34;</span>
</span></span><span class="line"><span class="cl">    <span class="n">output</span><span class="p">:</span> <span class="s2">&#34;out/points.csv&#34;</span>
</span></span><span class="line"><span class="cl">    <span class="n">run</span><span class="p">:</span> 
</span></span><span class="line"><span class="cl">        <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
</span></span><span class="line"><span class="cl">        <span class="n">df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;out/two_csv.csv&#39;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;temp_with_name.csv&#39;</span><span class="p">,</span><span class="s1">&#39;w&#39;</span><span class="p">)</span><span class="k">as</span> <span class="n">f</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">            <span class="n">name_df</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;scientific_name&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">            <span class="n">name_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">        <span class="n">df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;temp_with_name.csv&#39;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;points&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;scientific_name&#39;</span><span class="p">]</span><span class="o">/</span><span class="mi">5000</span><span class="o">*</span><span class="mi">100</span>
</span></span><span class="line"><span class="cl">        <span class="n">df</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">set_axis</span><span class="p">([</span><span class="s2">&#34;scientific_name&#34;</span><span class="p">,</span><span class="s2">&#34;hits&#34;</span><span class="p">,</span><span class="s2">&#34;points&#34;</span><span class="p">],</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&#34;out/points.csv&#34;</span><span class="p">,</span><span class="s2">&#34;w&#34;</span><span class="p">)</span><span class="k">as</span> <span class="n">f</span><span class="p">:</span>
</span></span><span class="line"><span class="cl">            <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="p">,</span><span class="n">float_format</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%.3f</span><span class="s1">&#39;</span><span class="p">,</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</span></span></code></pre></div>
        </div>
                
        
        

          
      </div>
     
      
      
        
        
          
          
        
      <script>
        var oid = "views_posts\/blast_nt\/index.md"
        var oid_likes = "likes_posts\/blast_nt\/index.md"
      </script>
      
      
      
      <script type="text/javascript" src="/js/page.min.b06a29d42a4ed16787978e2eee1e8c797b7698db2bc14ccee78f5c80ac566fc996190a73ad80a5e987558474b20b96fa38f7d85b405f165ff72b7b163c5ad11b.js" integrity="sha512-sGop1CpO0WeHl44u7h6MeXt2mNsrwUzO549cgKxWb8mWGQpzrYCl6YdVhHSyC5b6OPfYW0BfFl/3K3sWPFrRGw=="></script>
  
    </section>
  <footer class="pt-8 max-w-prose print:hidden">

    
  
    
    
    
    <div class="pt-8">
      <hr class="border-dotted border-neutral-300 dark:border-neutral-600" />
      <div class="flex justify-between pt-3">
        <span>
          
            <a class="flex group mr-3" href="/posts/zotero-cite/">
              <span
                class="mr-3 text-neutral-700 group-hover:text-primary-600 ltr:inline rtl:hidden dark:text-neutral dark:group-hover:text-primary-400"
                >&larr;</span
              >
              <span
                class="ml-3 text-neutral-700 group-hover:text-primary-600 ltr:hidden rtl:inline dark:text-neutral dark:group-hover:text-primary-400"
                >&rarr;</span
              >
              <span class="flex flex-col">
                <span
                  class="mt-[0.1rem] leading-6 group-hover:underline group-hover:decoration-primary-500"
                  >纯文本写作，随写随引</span
                >
                <span class="mt-[0.1rem] text-xs text-neutral-500 dark:text-neutral-400">
                  
                    <time datetime="2021-07-14 03:03:03 &#43;0800 CST">2021-07-14</time>
                  
                </span>
              </span>
            </a>
          
        </span>
        <span>
          
            <a class="flex text-right group ml-3" href="/posts/oneko/">
              <span class="flex flex-col">
                <span
                  class="mt-[0.1rem] leading-6 group-hover:underline group-hover:decoration-primary-500"
                  >oneko 一个小桌宠的折腾记录</span
                >
                <span class="mt-[0.1rem] text-xs text-neutral-500 dark:text-neutral-400">
                  
                    <time datetime="2021-11-10 00:00:00 &#43;0000 UTC">2021-11-10</time>
                  
                </span>
              </span>
              <span
                class="ml-3 text-neutral-700 group-hover:text-primary-600 ltr:inline rtl:hidden dark:text-neutral dark:group-hover:text-primary-400"
                >&rarr;</span
              >
              <span
                class="mr-3 text-neutral-700 group-hover:text-primary-600 ltr:hidden rtl:inline dark:text-neutral dark:group-hover:text-primary-400"
                >&larr;</span
              >
            </a>
          
        </span>
      </div>
    </div>
  


    
    
    <div class="pt-3">
      <hr class="border-dotted border-neutral-300 dark:border-neutral-600" />
      <div class="pt-3">
        <div id="gitalk-container"></div>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.css" crossorigin="anonymous">
    <script src="https://cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.js" crossorigin="anonymous"></script>
    <script type="text/javascript">
      var gitalk = new Gitalk({
        id: '2021-08-10 00:00:00 \u002b0000 UTC',
        title: '利用blast对测序数据进行污染情况估计',
        clientID: '27bf7867e6bf8c9bd64d',
        clientSecret: '9892a31bd3d6e5ca6e0f255ae3c4b2c761b33595',
        repo: 'with9.github.io',
        owner: 'with9',
        admin: ['with9'],
        body: decodeURI(location.href)
      });
      gitalk.render('gitalk-container');
    </script>
    <noscript>Please enable JavaScript to view the <a href="https://github.com/gitalk/gitalk">comments powered by gitalk.</a></noscript>

  

  

      </div>
    </div>
    
    
  </footer>
</article>

      <div id="top-scroller" class="pointer-events-none absolute top-[110vh] bottom-0 w-12 ltr:right-0 rtl:left-0">
  <a href="#the-top"
    class="pointer-events-auto sticky top-[calc(100vh-5.5rem)] flex h-12 w-12 mb-16 items-center justify-center rounded-full bg-neutral/50 text-xl text-neutral-700 hover:text-primary-600 dark:bg-neutral-800/50 dark:text-neutral dark:hover:text-primary-400"
    aria-label="Scroll to top" title="Scroll to top">
    &uarr;
  </a>
</div>
    </main><footer id="site-footer" class="py-10 print:hidden">
  
  
    
    <nav class="flex flex-row pb-4 text-base font-medium text-neutral-500 dark:text-neutral-400">
      <ul class="flex flex-col list-none sm:flex-row">
        
        <li class="flex mb-1 ltr:text-right rtl:text-left sm:mb-0 ltr:sm:mr-7 ltr:sm:last:mr-0 rtl:sm:ml-7 rtl:sm:last:ml-0">
          <a class="decoration-primary-500 hover:underline hover:decoration-2 hover:underline-offset-2 flex items-center" href="/tags/"
            title="">
            
            Tags
          </a>
        </li>
        
        <li class="flex mb-1 ltr:text-right rtl:text-left sm:mb-0 ltr:sm:mr-7 ltr:sm:last:mr-0 rtl:sm:ml-7 rtl:sm:last:ml-0">
          <a class="decoration-primary-500 hover:underline hover:decoration-2 hover:underline-offset-2 flex items-center" href="/authors/"
            title="">
            
            Authors
          </a>
        </li>
        
      </ul>
    </nav>
    
  
  <div class="flex items-center justify-between">

    
    
    <p class="text-sm text-neutral-500 dark:text-neutral-400">
      &copy;
      2024
      with Φ
    </p>
    

    
    
    <p class="text-xs text-neutral-500 dark:text-neutral-400">
      
      
      Powered by <a class="hover:underline hover:decoration-primary-400 hover:text-primary-500"
        href="https://gohugo.io/" target="_blank" rel="noopener noreferrer">Hugo</a> &amp; <a class="hover:underline hover:decoration-primary-400 hover:text-primary-500"
        href="https://blowfish.page/" target="_blank" rel="noopener noreferrer">Blowfish</a>
    </p>
    

  </div>
  <script>
    
    mediumZoom(document.querySelectorAll("img:not(.nozoom)"), {
      margin: 24,
      background: 'rgba(0,0,0,0.5)',
      scrollOffset: 0,
    })
    
  </script>
  
  
  <script type="text/javascript" src="/js/process.min.62060bb247f4de2b6dde45903668fefb68d792f365587605177b1227c0cf43588701edaca0cb40e2c8e2789bd5ce67c1d2a215b9fb258c3496a7cd25e7cb5fdf.js" integrity="sha512-YgYLskf03itt3kWQNmj&#43;&#43;2jXkvNlWHYFF3sSJ8DPQ1iHAe2soMtA4sjieJvVzmfB0qIVufsljDSWp80l58tf3w=="></script>
  
  
  <a rel="me" href="https://masto.ai/@blowfish"></a>
  
</footer><div
  id="search-wrapper"
  class="invisible fixed inset-0 flex h-screen w-screen cursor-default flex-col bg-neutral-500/50 p-4 backdrop-blur-sm dark:bg-neutral-900/50 sm:p-6 md:p-[10vh] lg:p-[12vh]"
  data-url="https://with9.github.io/"
  style="z-index:500"
>
  <div
    id="search-modal"
    class="flex flex-col w-full max-w-3xl min-h-0 mx-auto border rounded-md shadow-lg top-20 border-neutral-200 bg-neutral dark:border-neutral-700 dark:bg-neutral-800"
  >
    <header class="relative z-10 flex items-center justify-between flex-none px-2">
      <form class="flex items-center flex-auto min-w-0">
        <div class="flex items-center justify-center w-8 h-8 text-neutral-400">
          

  <span class="relative block icon">
    <svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="search" class="svg-inline--fa fa-search fa-w-16" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path fill="currentColor" d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"/></svg>

  </span>


        </div>
        <input
          type="search"
          id="search-query"
          class="flex flex-auto h-12 mx-1 bg-transparent appearance-none focus:outline-dotted focus:outline-2 focus:outline-transparent"
          placeholder="Search"
          tabindex="0"
        />
      </form>
      <button
        id="close-search-button"
        class="flex items-center justify-center w-8 h-8 text-neutral-700 hover:text-primary-600 dark:text-neutral dark:hover:text-primary-400"
        title="Close (Esc)"
      >
        

  <span class="relative block icon">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 320 512"><path fill="currentColor" d="M310.6 361.4c12.5 12.5 12.5 32.75 0 45.25C304.4 412.9 296.2 416 288 416s-16.38-3.125-22.62-9.375L160 301.3L54.63 406.6C48.38 412.9 40.19 416 32 416S15.63 412.9 9.375 406.6c-12.5-12.5-12.5-32.75 0-45.25l105.4-105.4L9.375 150.6c-12.5-12.5-12.5-32.75 0-45.25s32.75-12.5 45.25 0L160 210.8l105.4-105.4c12.5-12.5 32.75-12.5 45.25 0s12.5 32.75 0 45.25l-105.4 105.4L310.6 361.4z"/></svg>

  </span>


      </button>
    </header>
    <section class="flex-auto px-2 overflow-auto">
      <ul id="search-results">
        
      </ul>
    </section>
  </div>
</div>

  </div>
</body>

</html>
