<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
  <title>利用blast对测序数据进行污染情况估计 - 生活和睡觉.</title>
  <meta name="renderer" content="webkit" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>

<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />

<meta name="theme-color" content="#f8f5ec" />
<meta name="msapplication-navbutton-color" content="#f8f5ec">
<meta name="apple-mobile-web-app-capable" content="yes">
<meta name="apple-mobile-web-app-status-bar-style" content="#f8f5ec">


<meta name="author" content="with" /><meta name="description" content="一个简单的小流程,利用blast来统计测序数据的污染情况。" /><meta name="keywords" content="Hugo, theme, even" />






<meta name="generator" content="Hugo 0.86.0 with theme even" />


<link rel="canonical" href="https://with9.github.io/post/blast_nt/" />
<link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
<link rel="manifest" href="/manifest.json">
<link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5">

<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<link href="https://with9.github.io/sass/main.min.59fa635b67756aca024504fc61290d3c91e6332e773ecac116201a132e599ac0.css" rel="stylesheet">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@3.1.20/dist/jquery.fancybox.min.css" integrity="sha256-7TyXnr2YU040zfSP+rEcz29ggW4j56/ujTPwjMzyqFY=" crossorigin="anonymous">


<meta property="og:title" content="利用blast对测序数据进行污染情况估计" />
<meta property="og:description" content="一个简单的小流程,利用blast来统计测序数据的污染情况。" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://with9.github.io/post/blast_nt/" /><meta property="article:section" content="post" />
<meta property="article:published_time" content="2021-08-03T00:00:00+00:00" />
<meta property="article:modified_time" content="2021-08-03T00:00:00+00:00" />

<meta itemprop="name" content="利用blast对测序数据进行污染情况估计">
<meta itemprop="description" content="一个简单的小流程,利用blast来统计测序数据的污染情况。"><meta itemprop="datePublished" content="2021-08-03T00:00:00+00:00" />
<meta itemprop="dateModified" content="2021-08-03T00:00:00+00:00" />
<meta itemprop="wordCount" content="1367">
<meta itemprop="keywords" content="guide,Bio," /><meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="利用blast对测序数据进行污染情况估计"/>
<meta name="twitter:description" content="一个简单的小流程,利用blast来统计测序数据的污染情况。"/>

<!--[if lte IE 9]>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/classlist/1.1.20170427/classList.min.js"></script>
<![endif]-->

<!--[if lt IE 9]>
  <script src="https://cdn.jsdelivr.net/npm/html5shiv@3.7.3/dist/html5shiv.min.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/respond.js@1.4.2/dest/respond.min.js"></script>
<![endif]-->

</head>
<body>
  <div id="mobile-navbar" class="mobile-navbar">
  <div class="mobile-header-logo">
    <a href="/" class="logo">With&#39;s Blog</a>
  </div>
  <div class="mobile-navbar-icon">
    <span></span>
    <span></span>
    <span></span>
  </div>
</div>
<nav id="mobile-menu" class="mobile-menu slideout-menu">
  <ul class="mobile-menu-list">
    <a href="/">
        <li class="mobile-menu-item">Home</li>
      </a><a href="/post/">
        <li class="mobile-menu-item">Archives</li>
      </a><a href="/tags/">
        <li class="mobile-menu-item">Tags</li>
      </a><a href="/categories/">
        <li class="mobile-menu-item">Categories</li>
      </a><a href="/about/">
        <li class="mobile-menu-item">About</li>
      </a><a href="/friends/">
        <li class="mobile-menu-item">Friends</li>
      </a>
  </ul>
</nav>
  <div class="container" id="mobile-panel">
    <header id="header" class="header">
        <div class="logo-wrapper">
  <a href="/" class="logo">With&#39;s Blog</a>
</div>

<nav class="site-navbar">
  <ul id="menu" class="menu">
    <li class="menu-item">
        <a class="menu-item-link" href="/">Home</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/post/">Archives</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/tags/">Tags</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/categories/">Categories</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/about/">About</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/friends/">Friends</a>
      </li>
  </ul>
</nav>
    </header>

    <main id="main" class="main">
      <div class="content-wrapper">
        <div id="content" class="content">
          <article class="post">
    
    <header class="post-header">
      <h1 class="post-title">利用blast对测序数据进行污染情况估计</h1>

      <div class="post-meta">
        <span class="post-time"> 2021-08-03 </span>
        <div class="post-category">
            <a href="/categories/bio/"> Bio </a>
            </div>
          <span class="more-meta"> 1367 words </span>
          <span class="more-meta"> 3 mins read </span>
        <span id="busuanzi_container_page_pv" class="more-meta"> <span id="busuanzi_value_page_pv"><img src="/img/spinner.svg" alt="spinner.svg"/></span> times read </span>
      </div>
    </header>

    <div class="post-toc" id="post-toc">
  <h2 class="post-toc-title">Contents</h2>
  <div class="post-toc-content always-active">
    <nav id="TableOfContents">
  <ul>
    <li><a href="#前期数据库下载及预处理">前期数据库下载及预处理</a>
      <ul>
        <li><a href="#nt库下载">nt库下载</a></li>
        <li><a href="#物种名与基因id对应数据库下载">物种名与基因id对应数据库下载</a></li>
      </ul>
    </li>
    <li><a href="#对测序数据随机抽取并blast">对测序数据随机抽取并blast</a></li>
  </ul>
</nav>
  </div>
</div>
    <div class="post-content">
      <p>一般二代调研中需要判断一下测序样品的污染情况，主要的原理就是对样品的测序数据随机抽取并进行blast，分析reads的来源情况，从而判断样品污染情况。</p>
<p>主要流程参考<a href="https://blog.csdn.net/qq_42962326/article/details/105081327">这篇博客</a></p>
<h2 id="前期数据库下载及预处理">前期数据库下载及预处理</h2>
<h3 id="nt库下载">nt库下载</h3>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="c1">#nt.41.tar.gz</span>
<span class="c1">#https://ftp.ncbi.nlm.nih.gov/blast/db/nt.20.tar.gz</span>
<span class="c1">#----down.py</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">42</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">i</span> <span class="o">&lt;</span><span class="mi">10</span><span class="p">:</span>
        <span class="n">url</span><span class="o">=</span><span class="s2">&#34;https://ftp.ncbi.nlm.nih.gov/blast/db/nt.</span><span class="si">{}{}</span><span class="s2">.tar.gz&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;0&#39;</span><span class="p">,</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">url</span><span class="o">=</span><span class="s2">&#34;https://ftp.ncbi.nlm.nih.gov/blast/db/nt.</span><span class="si">{}</span><span class="s2">.tar.gz&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
</code></pre></td></tr></table>
</div>
</div><div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-shell" data-lang="shell">python down.py&gt;down
aria2c --input<span class="o">=</span>down <span class="c1">#进行下载</span>
<span class="c1">#分别解压</span>
</code></pre></td></tr></table>
</div>
</div><h3 id="物种名与基因id对应数据库下载">物种名与基因id对应数据库下载</h3>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span><span class="lnt">16
</span><span class="lnt">17
</span><span class="lnt">18
</span><span class="lnt">19
</span><span class="lnt">20
</span><span class="lnt">21
</span><span class="lnt">22
</span><span class="lnt">23
</span><span class="lnt">24
</span><span class="lnt">25
</span><span class="lnt">26
</span><span class="lnt">27
</span><span class="lnt">28
</span><span class="lnt">29
</span><span class="lnt">30
</span><span class="lnt">31
</span><span class="lnt">32
</span><span class="lnt">33
</span><span class="lnt">34
</span><span class="lnt">35
</span><span class="lnt">36
</span><span class="lnt">37
</span><span class="lnt">38
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-shell" data-lang="shell">aria2c <span class="s1">&#39;https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/accession2taxid/nucl_gb.accession2taxid.gz&#39;</span><span class="c1">#基因id对应taxonomy id</span>
aria2c <span class="s1">&#39;https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz&#39;</span><span class="c1">#taxonomy id 对应学名</span>
<span class="c1">#分别解压</span>
<span class="c1">#---head nucl_gb.accession2taxid</span>
accession	accession.version	taxid	gi
A00001	A00001.1	10641	<span class="m">58418</span>
A00002	A00002.1	9913	<span class="m">2</span>
A00003	A00003.1	9913	<span class="m">3</span>
A00004	A00004.1	32630	<span class="m">57971</span>
A00005	A00005.1	32630	<span class="m">57972</span>
A00006	A00006.1	32630	<span class="m">57973</span>
A00008	A00008.1	32630	<span class="m">57974</span>
A00009	A00009.1	32630	<span class="m">57975</span>
A00010	A00010.1	32630	<span class="m">57976</span>

<span class="c1">#---head taxdump/names.dmp</span>
1	<span class="p">|</span>	all	<span class="p">|</span>		<span class="p">|</span>	synonym	<span class="p">|</span>
1	<span class="p">|</span>	root	<span class="p">|</span>		<span class="p">|</span>	scientific name	<span class="p">|</span>
2	<span class="p">|</span>	Bacteria	<span class="p">|</span>	Bacteria &lt;bacteria&gt;	<span class="p">|</span>	scientific name	<span class="p">|</span>
2	<span class="p">|</span>	bacteria	<span class="p">|</span>		<span class="p">|</span>	blast name	<span class="p">|</span>
2	<span class="p">|</span>	eubacteria	<span class="p">|</span>		<span class="p">|</span>	genbank common name	<span class="p">|</span>
2	<span class="p">|</span>	Monera	<span class="p">|</span>	Monera &lt;bacteria&gt;	<span class="p">|</span>	in-part	<span class="p">|</span>
2	<span class="p">|</span>	Procaryotae	<span class="p">|</span>	Procaryotae &lt;bacteria&gt;	<span class="p">|</span>	in-part	<span class="p">|</span>
2	<span class="p">|</span>	Prokaryotae	<span class="p">|</span>	Prokaryotae &lt;bacteria&gt;	<span class="p">|</span>	in-part	<span class="p">|</span>
2	<span class="p">|</span>	Prokaryota	<span class="p">|</span>	Prokaryota &lt;bacteria&gt;	<span class="p">|</span>	in-part	<span class="p">|</span>
2	<span class="p">|</span>	prokaryote	<span class="p">|</span>	prokaryote &lt;bacteria&gt;	<span class="p">|</span>	in-part	<span class="p">|</span>
2	<span class="p">|</span>	prokaryotes	<span class="p">|</span>	prokaryotes &lt;bacteria&gt;	<span class="p">|</span>	in-part	<span class="p">|</span>
6	<span class="p">|</span>	Azorhizobium Dreyfus et al. <span class="m">1988</span> emend. Lang et al. 2013	<span class="p">|</span>		<span class="p">|</span>	authority	<span class="p">|</span>
6	<span class="p">|</span>	Azorhizobium	<span class="p">|</span>		<span class="p">|</span>	scientific name	<span class="p">|</span>
7	<span class="p">|</span>	ATCC 43989	<span class="p">|</span>	ATCC <span class="m">43989</span> &lt;<span class="nb">type</span> strain&gt;	<span class="p">|</span>	<span class="nb">type</span> material	<span class="p">|</span>
7	<span class="p">|</span>	Azorhizobium caulinodans Dreyfus et al. 1988	<span class="p">|</span>		<span class="p">|</span>	authority	<span class="p">|</span>
7	<span class="p">|</span>	Azorhizobium caulinodans	<span class="p">|</span>		<span class="p">|</span>	scientific name	<span class="p">|</span>
7	<span class="p">|</span>	Azotirhizobium caulinodans	<span class="p">|</span>		<span class="p">|</span>	equivalent name	<span class="p">|</span>
7	<span class="p">|</span>	CCUG 26647	<span class="p">|</span>	CCUG <span class="m">26647</span> &lt;<span class="nb">type</span> strain&gt;	<span class="p">|</span>	<span class="nb">type</span> material	<span class="p">|</span>
7	<span class="p">|</span>	DSM 5975	<span class="p">|</span>	DSM <span class="m">5975</span> &lt;<span class="nb">type</span> strain&gt;	<span class="p">|</span>	<span class="nb">type</span> material	<span class="p">|</span>
7	<span class="p">|</span>	IFO 14845	<span class="p">|</span>	IFO <span class="m">14845</span> &lt;<span class="nb">type</span> strain&gt;	<span class="p">|</span>	<span class="nb">type</span> material	<span class="p">|</span>

rga <span class="s2">&#34;scientific name&#34;</span> -g <span class="s2">&#34;names.dmp&#34;</span> &gt;names_scientific#获取scientific name相应的行
</code></pre></td></tr></table>
</div>
</div><h2 id="对测序数据随机抽取并blast">对测序数据随机抽取并blast</h2>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span><span class="lnt">16
</span><span class="lnt">17
</span><span class="lnt">18
</span><span class="lnt">19
</span><span class="lnt">20
</span><span class="lnt">21
</span><span class="lnt">22
</span><span class="lnt">23
</span><span class="lnt">24
</span><span class="lnt">25
</span><span class="lnt">26
</span><span class="lnt">27
</span><span class="lnt">28
</span><span class="lnt">29
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-shell" data-lang="shell"><span class="c1">#从fastq文件中抽取5000条数据</span>
cat 1.fastq <span class="p">|</span> head -n <span class="m">20000</span> <span class="p">|</span> awk <span class="s1">&#39;{if(NR%4==1){print &#34;&gt;&#34;$1}else if(NR%4==2){print $0}}&#39;</span><span class="p">|</span>sed <span class="s1">&#39;s/@//g&#39;</span> &gt; myfile.fa
<span class="c1">#---head myfile.fa</span>
&gt;SRR15225403.1
TTATCGTAGATTTTTTGAATATGTTGTAACGAAAGCTGAGCCATGTTGATTCCTTTGCTATACGTTTTGCCAGAAGGCCGATAACCGCGGCCAGGTGAGTGATTCCGTTGAAGGCAACAGTAATTGCGCCCCGGTTAAGCCCGCGCCGATCCCCACCGAGCGCATACCGCTGGCGTTAATGGCGTCAATGCCCGCCTGCGCATCTTCAATGCCGATACATGCCTGCGGCGGCAC
&gt;SRR15225403.2
GAAGGTTACTGCGGCTCCTGTCGCACCCGTCTGGTTGCAGGTCAGGTTGACTGGATTGCCGAACCGTTAGCCTTTATTCAGCCGGGGGAAATTTTGCCCTGTTGTTG
&gt;SRR15225403.3
CGGTAAACGCCGTCAGGGAGATGGCGCAACCAATCGCCAACGGCAGATTAGCCCACAGACCCATCACGATAGAACCGAGTCCGGCAACCAGACAGGTCGCAACGAAAACTGCCGCAGGCGGGAAGCCCGCTTTACCCAACATGCCTGGAACGACGATAACCGAGTAGACCATCGCCAGAAACGTTGTTAACCCGGCAACCACTTCCTGACGGACAGTGCTTCCACGTTGTGAAAT
&gt;SRR15225403.4
CAGGCGAAAGCCGCGGCGGTAAAAGTCCACGCTGACGCGCCCGGTACGTTTTATTGCGGATGTAAAATTAACTGGCAGGGCAAAAAAGGCGTTGTTGATCTGCAATCGTGCGGCTATCAGGTGCGCAAAAATGAAAACCGAGCCAGCCGCGTAGAGTGGGAACACGTCGTTCCCGCCTGGCAGTTCGGTCACCAGCGCCAGTGCTGGCAGGACGGGGGACGTAAAAACTGCGCTA
&gt;SRR15225403.5
TGTATCTGAATCCGCAAGATTGCAGCGTGATTAATGATGAAGCGCTGAATCGTATTATCGCCGTAGGCCACCAGCATCATCTGAACGTTGTCGGCTGGAACCCGGGACCGGCGCTTTCAGTTAGCATGGGCGACATGCCGGATGATGGCTACAAACCATTGGTTGGGTAGAAAAGGCCTACGGCTTCGGAACGGCAAAAGGGAACAAAAGGGAAACTGCCACACCTGGCGAATC

<span class="c1">#进行blastn 并将结果保存为xml格式</span>
blastn -query myfile.fa -out xml/out00.xml -max_target_seqs <span class="m">1</span> -outfmt <span class="m">5</span> -db ~/usb/blastn/nt.00/nt.00 -num_threads <span class="m">2</span> -evalue 1e-5

<span class="c1">#---head out00.xml</span>
&lt;?xml <span class="nv">version</span><span class="o">=</span><span class="s2">&#34;1.0&#34;</span>?&gt;
&lt;!DOCTYPE BlastOutput PUBLIC <span class="s2">&#34;-//NCBI//NCBI BlastOutput/EN&#34;</span> <span class="s2">&#34;http://www.ncbi.nlm.nih.gov/dtd/NCBI_BlastOutput.dtd&#34;</span>&gt;
&lt;BlastOutput&gt;
  &lt;BlastOutput_program&gt;blastn&lt;/BlastOutput_program&gt;
  &lt;BlastOutput_version&gt;BLASTN 2.12.0+&lt;/BlastOutput_version&gt;
  &lt;BlastOutput_reference&gt;Zheng Zhang, Scott Schwartz, Lukas Wagner, and Webb Miller <span class="o">(</span>2000<span class="o">)</span>, <span class="p">&amp;</span>quot<span class="p">;</span>A greedy algorithm <span class="k">for</span> aligning DNA sequences<span class="p">&amp;</span>quot<span class="p">;</span>, J Comput Biol 2000<span class="p">;</span> 7<span class="o">(</span>1-2<span class="o">)</span>:203-14.&lt;/BlastOutput_reference&gt;
  &lt;BlastOutput_db&gt;/home/with9/usb/blastn/nt.00/nt.00&lt;/BlastOutput_db&gt;
  &lt;BlastOutput_query-ID&gt;Query_1&lt;/BlastOutput_query-ID&gt;
  &lt;BlastOutput_query-def&gt;SRR15225403.1&lt;/BlastOutput_query-def&gt;
  &lt;BlastOutput_query-len&gt;234&lt;/BlastOutput_query-len&gt;

</code></pre></td></tr></table>
</div>
</div><p>编写python脚本处理out.xml文件,将gid,hit_def和hit_accession写入csv文件</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span><span class="lnt">16
</span><span class="lnt">17
</span><span class="lnt">18
</span><span class="lnt">19
</span><span class="lnt">20
</span><span class="lnt">21
</span><span class="lnt">22
</span><span class="lnt">23
</span><span class="lnt">24
</span><span class="lnt">25
</span><span class="lnt">26
</span><span class="lnt">27
</span><span class="lnt">28
</span><span class="lnt">29
</span><span class="lnt">30
</span><span class="lnt">31
</span><span class="lnt">32
</span><span class="lnt">33
</span><span class="lnt">34
</span><span class="lnt">35
</span><span class="lnt">36
</span><span class="lnt">37
</span><span class="lnt">38
</span><span class="lnt">39
</span><span class="lnt">40
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="kn">import</span> <span class="nn">xml.etree.ElementTree</span> <span class="k">as</span> <span class="nn">ET</span>
<span class="kn">import</span> <span class="nn">csv</span>
<span class="kn">import</span> <span class="nn">os</span>

<span class="k">def</span> <span class="nf">readonexml</span><span class="p">(</span><span class="n">fid</span><span class="p">:</span><span class="nb">int</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">fid</span><span class="o">&lt;</span><span class="mi">10</span><span class="p">:</span>
        <span class="n">xmlfile</span><span class="o">=</span><span class="s2">&#34;xml/out0</span><span class="si">{}</span><span class="s2">.xml&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">fid</span><span class="p">)</span>
        <span class="n">csvfile</span><span class="o">=</span><span class="s2">&#34;csv/out0</span><span class="si">{}</span><span class="s2">.csv&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">fid</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">xmlfile</span><span class="o">=</span><span class="s2">&#34;xml/out</span><span class="si">{}</span><span class="s2">.xml&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">fid</span><span class="p">)</span>
        <span class="n">csvfile</span><span class="o">=</span><span class="s2">&#34;csv/out</span><span class="si">{}</span><span class="s2">.csv&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">fid</span><span class="p">)</span>

    <span class="n">tree</span> <span class="o">=</span> <span class="n">ET</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">xmlfile</span><span class="p">)</span>
    <span class="n">root</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">getroot</span><span class="p">()</span>
    <span class="n">hits</span><span class="o">=</span><span class="n">root</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="s1">&#39;.//Hit&#39;</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">csvfile</span><span class="p">,</span><span class="s1">&#39;w&#39;</span><span class="p">,</span><span class="n">newline</span><span class="o">=</span><span class="s2">&#34;&#34;</span><span class="p">)</span><span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">csv_writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
        <span class="n">csv_writer</span><span class="o">.</span><span class="n">writerow</span><span class="p">([</span><span class="s2">&#34;Hit_id&#34;</span><span class="p">,</span><span class="s2">&#34;Hit_def&#34;</span><span class="p">,</span><span class="s2">&#34;Hit_accession&#34;</span><span class="p">])</span>

        <span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">hits</span><span class="p">:</span>
            <span class="n">hit_id</span><span class="o">=</span><span class="n">hit</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;Hit_id&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>
            <span class="n">hit_def</span><span class="o">=</span><span class="n">hit</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;Hit_def&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>
            <span class="n">hit_accession</span><span class="o">=</span><span class="n">hit</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;Hit_accession&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>
            <span class="n">csv_writer</span><span class="o">.</span><span class="n">writerow</span><span class="p">([</span><span class="n">hit_id</span><span class="p">,</span><span class="n">hit_def</span><span class="p">,</span><span class="n">hit_accession</span><span class="p">])</span>

<span class="n">readonexml</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

<span class="c1">#--head csv/out00.csv</span>
<span class="n">Hit_id</span><span class="p">,</span><span class="n">Hit_def</span><span class="p">,</span><span class="n">Hit_accession</span>
<span class="n">gi</span><span class="o">|</span><span class="mi">85674274</span><span class="o">|</span><span class="n">dbj</span><span class="o">|</span><span class="n">AP009048</span><span class="mf">.1</span><span class="o">|</span><span class="p">,</span><span class="s2">&#34;Escherichia coli str. K-12 substr. W3110 DNA, complete genome&#34;</span><span class="p">,</span><span class="n">AP009048</span>
<span class="n">gi</span><span class="o">|</span><span class="mi">85674274</span><span class="o">|</span><span class="n">dbj</span><span class="o">|</span><span class="n">AP009048</span><span class="mf">.1</span><span class="o">|</span><span class="p">,</span><span class="s2">&#34;Escherichia coli str. K-12 substr. W3110 DNA, complete genome&#34;</span><span class="p">,</span><span class="n">AP009048</span>
<span class="n">gi</span><span class="o">|</span><span class="mi">81239530</span><span class="o">|</span><span class="n">gb</span><span class="o">|</span><span class="n">CP000034</span><span class="mf">.1</span><span class="o">|</span><span class="p">,</span><span class="s2">&#34;Shigella dysenteriae Sd197, complete genome&#34;</span><span class="p">,</span><span class="n">CP000034</span>
<span class="n">gi</span><span class="o">|</span><span class="mi">81244029</span><span class="o">|</span><span class="n">gb</span><span class="o">|</span><span class="n">CP000036</span><span class="mf">.1</span><span class="o">|</span><span class="p">,</span><span class="s2">&#34;Shigella boydii Sb227, complete genome&#34;</span><span class="p">,</span><span class="n">CP000036</span>
<span class="n">gi</span><span class="o">|</span><span class="mi">30043918</span><span class="o">|</span><span class="n">gb</span><span class="o">|</span><span class="n">AE014073</span><span class="mf">.1</span><span class="o">|</span><span class="p">,</span><span class="s2">&#34;Shigella flexneri 2a str. 2457T, complete genome&#34;</span><span class="p">,</span><span class="n">AE014073</span>
<span class="n">gi</span><span class="o">|</span><span class="mi">30043918</span><span class="o">|</span><span class="n">gb</span><span class="o">|</span><span class="n">AE014073</span><span class="mf">.1</span><span class="o">|</span><span class="p">,</span><span class="s2">&#34;Shigella flexneri 2a str. 2457T, complete genome&#34;</span><span class="p">,</span><span class="n">AE014073</span>
<span class="n">gi</span><span class="o">|</span><span class="mi">81244029</span><span class="o">|</span><span class="n">gb</span><span class="o">|</span><span class="n">CP000036</span><span class="mf">.1</span><span class="o">|</span><span class="p">,</span><span class="s2">&#34;Shigella boydii Sb227, complete genome&#34;</span><span class="p">,</span><span class="n">CP000036</span>
<span class="n">gi</span><span class="o">|</span><span class="mi">85674274</span><span class="o">|</span><span class="n">dbj</span><span class="o">|</span><span class="n">AP009048</span><span class="mf">.1</span><span class="o">|</span><span class="p">,</span><span class="s2">&#34;Escherichia coli str. K-12 substr. W3110 DNA, complete genome&#34;</span><span class="p">,</span><span class="n">AP009048</span>
<span class="n">gi</span><span class="o">|</span><span class="mi">30043918</span><span class="o">|</span><span class="n">gb</span><span class="o">|</span><span class="n">AE014073</span><span class="mf">.1</span><span class="o">|</span><span class="p">,</span><span class="s2">&#34;Shigella flexneri 2a str. 2457T, complete genome&#34;</span><span class="p">,</span><span class="n">AE014073</span>

</code></pre></td></tr></table>
</div>
</div><p>利用Hit_accession从nucl_gb.accession2taxid中获取taxonomyid,因为文件过大,所以利用os模块调用<a href="https://github.com/phiresky/ripgrep-all">rga</a>工具进行查找</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span><span class="lnt">16
</span><span class="lnt">17
</span><span class="lnt">18
</span><span class="lnt">19
</span><span class="lnt">20
</span><span class="lnt">21
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="c1">#统计出所有需要查找的Hit_accession,因为文件过大,所以利用os模块调用rga工具进行查找</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="n">df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;csv/out00.csv&#39;</span><span class="p">)</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;Hitaccession.csv&#39;</span><span class="p">,</span><span class="s1">&#39;w&#39;</span><span class="p">)</span><span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="n">name_df</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Hit_accession&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
    <span class="n">name_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
    

<span class="n">df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;Hitacession.csv&#39;</span><span class="p">)</span>
<span class="n">df</span><span class="p">[</span><span class="s1">&#39;tax_id&#39;</span><span class="p">]</span><span class="o">=</span><span class="mi">0</span>
<span class="n">tids</span><span class="o">=</span><span class="p">[]</span>
<span class="k">for</span> <span class="n">index</span><span class="p">,</span><span class="n">row</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="n">hit_accession</span><span class="o">=</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">cmd</span><span class="o">=</span><span class="s2">&#34;rga &#39;&#34;</span><span class="o">+</span><span class="n">hit_accession</span><span class="o">+</span><span class="s2">&#34;&#39; -g &#39;nucl_gb.accession2taxid.gz&#39;|awk &#39;{print$3}&#39;&#34;</span>
    <span class="n">tid</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">popen</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">tid</span><span class="p">)</span>
    <span class="n">tids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tid</span><span class="p">)</span>

<span class="n">df</span><span class="p">[</span><span class="s1">&#39;tax_id&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">tids</span>
<span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;Tid_result,csv&#39;</span><span class="p">)</span>
</code></pre></td></tr></table>
</div>
</div><p>对一开始的读取xml脚本进行修改,同时获取物种id和物种的学名.</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span><span class="lnt">16
</span><span class="lnt">17
</span><span class="lnt">18
</span><span class="lnt">19
</span><span class="lnt">20
</span><span class="lnt">21
</span><span class="lnt">22
</span><span class="lnt">23
</span><span class="lnt">24
</span><span class="lnt">25
</span><span class="lnt">26
</span><span class="lnt">27
</span><span class="lnt">28
</span><span class="lnt">29
</span><span class="lnt">30
</span><span class="lnt">31
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="kn">import</span> <span class="nn">xml.etree.ElementTree</span> <span class="k">as</span> <span class="nn">ET</span>
<span class="kn">import</span> <span class="nn">csv</span>
<span class="kn">import</span> <span class="nn">os</span>

<span class="k">def</span> <span class="nf">readonexml</span><span class="p">(</span><span class="n">fid</span><span class="p">:</span><span class="nb">int</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">fid</span><span class="o">&lt;</span><span class="mi">10</span><span class="p">:</span>
        <span class="n">xmlfile</span><span class="o">=</span><span class="s2">&#34;xml/out0</span><span class="si">{}</span><span class="s2">.xml&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">fid</span><span class="p">)</span>
        <span class="n">csvfile</span><span class="o">=</span><span class="s2">&#34;csv/out0</span><span class="si">{}</span><span class="s2">.csv&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">fid</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">xmlfile</span><span class="o">=</span><span class="s2">&#34;xml/out</span><span class="si">{}</span><span class="s2">.xml&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">fid</span><span class="p">)</span>
        <span class="n">csvfile</span><span class="o">=</span><span class="s2">&#34;csv/out</span><span class="si">{}</span><span class="s2">.csv&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">fid</span><span class="p">)</span>

    <span class="n">tree</span> <span class="o">=</span> <span class="n">ET</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">xmlfile</span><span class="p">)</span>
    <span class="n">root</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">getroot</span><span class="p">()</span>
    <span class="n">hits</span><span class="o">=</span><span class="n">root</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="s1">&#39;.//Hit&#39;</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">csvfile</span><span class="p">,</span><span class="s1">&#39;w&#39;</span><span class="p">,</span><span class="n">newline</span><span class="o">=</span><span class="s2">&#34;&#34;</span><span class="p">)</span><span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">csv_writer</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
        <span class="n">csv_writer</span><span class="o">.</span><span class="n">writerow</span><span class="p">([</span><span class="s2">&#34;Hit_id&#34;</span><span class="p">,</span><span class="s2">&#34;Hit_def&#34;</span><span class="p">,</span><span class="s2">&#34;Hit_accession&#34;</span><span class="p">,</span><span class="s2">&#34;taxonomy_id&#34;</span><span class="p">,</span><span class="s2">&#34;scientific_name&#34;</span><span class="p">])</span>

        <span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">hits</span><span class="p">:</span>
            <span class="n">hit_id</span><span class="o">=</span><span class="n">hit</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;Hit_id&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>
            <span class="n">hit_def</span><span class="o">=</span><span class="n">hit</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;Hit_def&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>
            <span class="n">hit_accession</span><span class="o">=</span><span class="n">hit</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s1">&#39;Hit_accession&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">text</span>
            <span class="n">cmd</span><span class="o">=</span><span class="s2">&#34;grep &#39;</span><span class="si">{}</span><span class="s2">&#39; Tid_result.csv |cut -d , -f 4&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">hit_accession</span><span class="p">)</span>
            <span class="n">taxonomy_id</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">popen</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">taxonomy_id</span><span class="p">)</span>
            <span class="n">cmd2</span><span class="o">=</span><span class="s2">&#34;grep &#39;:</span><span class="si">{}</span><span class="s2">&#39; names_scientific |cut -d \| -f 2&#34;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">taxonomy_id</span><span class="p">)</span>
            <span class="n">scientific_name</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">popen</span><span class="p">(</span><span class="n">cmd2</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">csv_writer</span><span class="o">.</span><span class="n">writerow</span><span class="p">([</span><span class="n">hit_id</span><span class="p">,</span><span class="n">hit_def</span><span class="p">,</span><span class="n">hit_accession</span><span class="p">,</span><span class="n">taxonomy_id</span><span class="p">,</span><span class="n">scientific_name</span><span class="p">])</span>
<span class="n">readonexml</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
</code></pre></td></tr></table>
</div>
</div><p>统计结果并绘图</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span><span class="lnt">16
</span><span class="lnt">17
</span><span class="lnt">18
</span><span class="lnt">19
</span><span class="lnt">20
</span><span class="lnt">21
</span><span class="lnt">22
</span><span class="lnt">23
</span><span class="lnt">24
</span><span class="lnt">25
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-python" data-lang="python"><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;csv/out00.csv&#39;</span><span class="p">)</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;temp_with_name.csv&#39;</span><span class="p">,</span><span class="s1">&#39;w&#39;</span><span class="p">)</span><span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="n">name_df</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;scientific_name&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
    <span class="n">name_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>

<span class="n">name_df</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">pie</span><span class="p">()</span>
<span class="n">df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;temp_with_name.csv&#39;</span><span class="p">)</span>
<span class="n">df</span><span class="p">[</span><span class="s1">&#39;points&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;scientific_name&#39;</span><span class="p">]</span><span class="o">/</span><span class="mi">5000</span><span class="o">*</span><span class="mi">100</span>
<span class="n">df</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">set_axis</span><span class="p">([</span><span class="s2">&#34;scientific_name&#34;</span><span class="p">,</span><span class="s2">&#34;hits&#34;</span><span class="p">,</span><span class="s2">&#34;points&#34;</span><span class="p">],</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&#34;point.csv&#34;</span><span class="p">,</span><span class="s2">&#34;w&#34;</span><span class="p">)</span><span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="p">,</span><span class="n">float_format</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%.3f</span><span class="s1">&#39;</span><span class="p">,</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="c1">#---head point.csv</span>
<span class="n">scientific_name</span><span class="p">,</span><span class="n">hits</span><span class="p">,</span><span class="n">points</span>
<span class="n">Escherichia</span> <span class="n">coli</span> <span class="nb">str</span><span class="o">.</span> <span class="n">K</span><span class="o">-</span><span class="mi">12</span> <span class="n">substr</span><span class="o">.</span> <span class="n">W3110</span><span class="p">,</span><span class="mi">1785</span><span class="p">,</span><span class="mf">35.700</span>
<span class="n">Shigella</span> <span class="n">boydii</span> <span class="n">Sb227</span><span class="p">,</span><span class="mi">668</span><span class="p">,</span><span class="mf">13.360</span>
<span class="n">Shigella</span> <span class="n">flexneri</span> <span class="mi">2</span><span class="n">a</span> <span class="nb">str</span><span class="o">.</span> <span class="mi">2457</span><span class="n">T</span><span class="p">,</span><span class="mi">614</span><span class="p">,</span><span class="mf">12.280</span>
<span class="n">Shigella</span> <span class="n">sonnei</span> <span class="n">Ss046</span><span class="p">,</span><span class="mi">438</span><span class="p">,</span><span class="mf">8.760</span>
<span class="n">Escherichia</span> <span class="n">coli</span><span class="p">,</span><span class="mi">244</span><span class="p">,</span><span class="mf">4.880</span>
<span class="n">Escherichia</span> <span class="n">coli</span> <span class="n">CFT073</span><span class="p">,</span><span class="mi">242</span><span class="p">,</span><span class="mf">4.840</span>
<span class="n">Escherichia</span> <span class="n">coli</span> <span class="n">O157</span><span class="p">:</span><span class="n">H7</span> <span class="nb">str</span><span class="o">.</span> <span class="n">EDL933</span><span class="p">,</span><span class="mi">233</span><span class="p">,</span><span class="mf">4.660</span>
<span class="n">Shigella</span> <span class="n">dysenteriae</span> <span class="n">Sd197</span><span class="p">,</span><span class="mi">149</span><span class="p">,</span><span class="mf">2.980</span>
<span class="n">Escherichia</span> <span class="n">coli</span> <span class="n">K</span><span class="o">-</span><span class="mi">12</span><span class="p">,</span><span class="mi">55</span><span class="p">,</span><span class="mf">1.100</span>

</code></pre></td></tr></table>
</div>
</div>
    </div>

    <div class="post-copyright">
  <p class="copyright-item">
    <span class="item-title">Author</span>
    <span class="item-content">with</span>
  </p>
  <p class="copyright-item">
    <span class="item-title">LastMod</span>
    <span class="item-content">
        2021-08-03
        
    </span>
  </p>
  <p class="copyright-item">
      <span class="item-title">Markdown</span>
      <span class="item-content"><a class="link-to-markdown" href="https://with9.github.io/post/blast_nt/index.md" target="_blank">The Markdown version »</a></span>
    </p>
  
</div>
<footer class="post-footer">
      <div class="post-tags">
          <a href="/tags/guide/">guide</a>
          <a href="/tags/bio/">Bio</a>
          </div>
      <nav class="post-nav">
        
        <a class="next" href="/post/zotero-cite/">
            <span class="next-text nav-default">纯文本写作，随写随引</span>
            <span class="next-text nav-mobile">Next</span>
            <i class="iconfont icon-right"></i>
          </a>
      </nav>
    </footer>
  </article>
        </div>
        <div id="gitalk-container"></div>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.css" crossorigin="anonymous">
    <script src="https://cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.js" crossorigin="anonymous"></script>
    <script type="text/javascript">
      var gitalk = new Gitalk({
        id: '2021-08-03 00:00:00 \u002b0000 UTC',
        title: '利用blast对测序数据进行污染情况估计',
        clientID: '27bf7867e6bf8c9bd64d',
        clientSecret: '9892a31bd3d6e5ca6e0f255ae3c4b2c761b33595',
        repo: 'with9.github.io',
        owner: 'with9',
        admin: ['with9'],
        body: decodeURI(location.href)
      });
      gitalk.render('gitalk-container');
    </script>
    <noscript>Please enable JavaScript to view the <a href="https://github.com/gitalk/gitalk">comments powered by gitalk.</a></noscript>

  

  

      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="social-links">
      <a href="mailto:dc198424601@outlook.com" class="iconfont icon-email" title="email"></a>
      <a href="https://github.com/with9" class="iconfont icon-github" title="github"></a>
      <a href="https://weibo.com/u/2582158737" class="iconfont icon-weibo" title="weibo"></a>
      <a href="https://www.zhihu.com/people/yu-yuan-86-34" class="iconfont icon-zhihu" title="zhihu"></a>
      <a href="https://space.bilibili.com/9063101" class="iconfont icon-bilibili" title="bilibili"></a>
  <a href="https://with9.github.io/index.xml" type="application/rss+xml" class="iconfont icon-rss" title="rss"></a>
</div>

<div class="copyright">
  <span class="power-by">
    Powered by <a class="hexo-link" href="https://gohugo.io">Hugo</a>
  </span>
  <span class="division">|</span>
  <span class="theme-info">
    Theme - 
    <a class="theme-link" href="https://github.com/olOwOlo/hugo-theme-even">Even</a>
  </span>

  <div class="busuanzi-footer">
    <span id="busuanzi_container_site_pv"> site pv: <span id="busuanzi_value_site_pv"><img src="/img/spinner.svg" alt="spinner.svg"/></span> </span>
      <span class="division">|</span>
    <span id="busuanzi_container_site_uv"> site uv: <span id="busuanzi_value_site_uv"><img src="/img/spinner.svg" alt="spinner.svg"/></span> </span>
  </div>

  <span class="copyright-year">
    &copy; 
    2020 - 
    2021
    <span class="heart">
      <i class="iconfont icon-heart"></i>
    </span>
    <span class="author">with</span>
  </span>
</div>
    </footer>

    <div class="back-to-top" id="back-to-top">
      <i class="iconfont icon-up"></i>
    </div>
  </div>
  
  <script src="https://cdn.jsdelivr.net/npm/jquery@3.2.1/dist/jquery.min.js" integrity="sha256-hwg4gsxgFZhOsEEamdOYGBf13FyQuiTwlAQgxVSNgt4=" crossorigin="anonymous"></script>
  <script src="https://cdn.jsdelivr.net/npm/slideout@1.0.1/dist/slideout.min.js" integrity="sha256-t+zJ/g8/KXIJMjSVQdnibt4dlaDxc9zXr/9oNPeWqdg=" crossorigin="anonymous"></script>
  <script src="https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@3.1.20/dist/jquery.fancybox.min.js" integrity="sha256-XVLffZaxoWfGUEbdzuLi7pwaUJv1cecsQJQqGLe7axY=" crossorigin="anonymous"></script>



<script type="text/javascript" src="https://with9.github.io/js/main.4d21c1cefdf1cd0b77ef59bab2d1bbe3112fc9d6cc04710bb1f0b1fd7b891f11.js"></script>








</body>
</html>
